{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEBUG = False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OrthoGroups to MSA \n",
    "+ load table with the proteins - orthogrous distribution \n",
    "+ filter the core orthogroups \n",
    "+ load proteins .pkl database\n",
    "+ get the proteins and the genome positions corresponding to the core OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_ogs = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_rso_test/Species_Tree/Orthogroups_for_concatenated_alignment.txt'\n",
    "ogs2prot = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_rso_test/Orthogroups/Orthogroups.tsv'\n",
    "prot_db = pd.read_pickle(\"../outputs/GENOME_PROTEIN_DB_rso_test.pkl\")\n",
    "path_p_aln = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_rso_test/MultipleSequenceAlignments/'\n",
    "\n",
    "msa_out_fasta = \"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/rso_test_concat.fasta\"\n",
    "msa_out_phylip = \"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/rso_test_concat.phylip\"\n",
    "msa_out_xmfa = \"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/rso_test_concat.xmfa\"\n",
    "aln2acc = \"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/outputs/MSA_ACC_DB_rso_test.pkl\"\n",
    "\n",
    "#core_ogs = snakemake.input[\"core_ogs\"]\n",
    "#ogs2prot = snakemake.input[\"ogs2prot\"]\n",
    "#prot_db = pd.read_pickle(snakemake.input[\"prot_db\"])\n",
    "#path_p_aln = snakemake.input[\"path2aln\"]\n",
    "\n",
    "#msa_out_fasta = snakemake.output[\"msa_fasta\"]\n",
    "#msa_out_phylip = snakemake.output[\"msa_phylip\"]\n",
    "#msa_out_xmfa = snakemake.output[\"msa_xmfa\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get core ogs\n",
    "f = open(core_ogs,\"r\")\n",
    "ogs = []\n",
    "for l in f:\n",
    "    ogs.append(l.replace('\\n',''))\n",
    "f.close()\n",
    "##print(ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_001887535.1_ASM188753v1_protein</th>\n",
       "      <th>GCF_013306235.1_ASM1330623v1_protein</th>\n",
       "      <th>GCF_013306335.1_ASM1330633v1_protein</th>\n",
       "      <th>GCF_013306435.1_ASM1330643v1_protein</th>\n",
       "      <th>GCF_013306935.1_ASM1330693v1_protein</th>\n",
       "      <th>GCF_013375735.1_ASM1337573v1_protein</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orthogroup</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OG0000225</th>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000226</th>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000227</th>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000228</th>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000229</th>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004089</th>\n",
       "      <td>WP_155772886.1</td>\n",
       "      <td>WP_148668905.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_155772886.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004091</th>\n",
       "      <td>WP_155773184.1</td>\n",
       "      <td>WP_173940865.1</td>\n",
       "      <td>WP_173941733.1</td>\n",
       "      <td>WP_119889930.1</td>\n",
       "      <td>WP_173952290.1</td>\n",
       "      <td>WP_086706436.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004092</th>\n",
       "      <td>WP_155773186.1</td>\n",
       "      <td>WP_144061950.1</td>\n",
       "      <td>WP_119417730.1</td>\n",
       "      <td>WP_119889926.1</td>\n",
       "      <td>WP_119417730.1</td>\n",
       "      <td>WP_038938529.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004093</th>\n",
       "      <td>WP_155773188.1</td>\n",
       "      <td>WP_155739038.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155773188.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004094</th>\n",
       "      <td>WP_155773189.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "      <td>WP_058907847.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "      <td>WP_058907847.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3842 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GCF_001887535.1_ASM188753v1_protein  \\\n",
       "Orthogroup                                       \n",
       "OG0000225                       WP_003261911.1   \n",
       "OG0000226                       WP_003261929.1   \n",
       "OG0000227                       WP_003261939.1   \n",
       "OG0000228                       WP_003261941.1   \n",
       "OG0000229                       WP_003262123.1   \n",
       "...                                        ...   \n",
       "OG0004089                       WP_155772886.1   \n",
       "OG0004091                       WP_155773184.1   \n",
       "OG0004092                       WP_155773186.1   \n",
       "OG0004093                       WP_155773188.1   \n",
       "OG0004094                       WP_155773189.1   \n",
       "\n",
       "           GCF_013306235.1_ASM1330623v1_protein  \\\n",
       "Orthogroup                                        \n",
       "OG0000225                        WP_003261911.1   \n",
       "OG0000226                        WP_003261929.1   \n",
       "OG0000227                        WP_003261939.1   \n",
       "OG0000228                        WP_003261941.1   \n",
       "OG0000229                        WP_003262123.1   \n",
       "...                                         ...   \n",
       "OG0004089                        WP_148668905.1   \n",
       "OG0004091                        WP_173940865.1   \n",
       "OG0004092                        WP_144061950.1   \n",
       "OG0004093                        WP_155739038.1   \n",
       "OG0004094                        WP_021155598.1   \n",
       "\n",
       "           GCF_013306335.1_ASM1330633v1_protein  \\\n",
       "Orthogroup                                        \n",
       "OG0000225                        WP_003261911.1   \n",
       "OG0000226                        WP_003261929.1   \n",
       "OG0000227                        WP_003261939.1   \n",
       "OG0000228                        WP_003261941.1   \n",
       "OG0000229                        WP_003262123.1   \n",
       "...                                         ...   \n",
       "OG0004089                        WP_148668915.1   \n",
       "OG0004091                        WP_173941733.1   \n",
       "OG0004092                        WP_119417730.1   \n",
       "OG0004093                        WP_155772892.1   \n",
       "OG0004094                        WP_058907847.1   \n",
       "\n",
       "           GCF_013306435.1_ASM1330643v1_protein  \\\n",
       "Orthogroup                                        \n",
       "OG0000225                        WP_003261911.1   \n",
       "OG0000226                        WP_003261929.1   \n",
       "OG0000227                        WP_003261939.1   \n",
       "OG0000228                        WP_003261941.1   \n",
       "OG0000229                        WP_003262123.1   \n",
       "...                                         ...   \n",
       "OG0004089                        WP_148668915.1   \n",
       "OG0004091                        WP_119889930.1   \n",
       "OG0004092                        WP_119889926.1   \n",
       "OG0004093                        WP_155772892.1   \n",
       "OG0004094                        WP_021155598.1   \n",
       "\n",
       "           GCF_013306935.1_ASM1330693v1_protein  \\\n",
       "Orthogroup                                        \n",
       "OG0000225                        WP_003261911.1   \n",
       "OG0000226                        WP_003261929.1   \n",
       "OG0000227                        WP_003261939.1   \n",
       "OG0000228                        WP_003261941.1   \n",
       "OG0000229                        WP_003262123.1   \n",
       "...                                         ...   \n",
       "OG0004089                        WP_148668915.1   \n",
       "OG0004091                        WP_173952290.1   \n",
       "OG0004092                        WP_119417730.1   \n",
       "OG0004093                        WP_155772892.1   \n",
       "OG0004094                        WP_058907847.1   \n",
       "\n",
       "           GCF_013375735.1_ASM1337573v1_protein  \n",
       "Orthogroup                                       \n",
       "OG0000225                        WP_003261911.1  \n",
       "OG0000226                        WP_003261929.1  \n",
       "OG0000227                        WP_003261939.1  \n",
       "OG0000228                        WP_003261941.1  \n",
       "OG0000229                        WP_003262123.1  \n",
       "...                                         ...  \n",
       "OG0004089                        WP_155772886.1  \n",
       "OG0004091                        WP_086706436.1  \n",
       "OG0004092                        WP_038938529.1  \n",
       "OG0004093                        WP_155773188.1  \n",
       "OG0004094                        WP_021155598.1  \n",
       "\n",
       "[3842 rows x 6 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load roteins ids from the OGs table \n",
    "# keep only core the OGs\n",
    "\n",
    "tab_prot = pd.read_csv(ogs2prot, sep='\\t', index_col=0)  \n",
    "tab_prot = tab_prot.loc[ogs]\n",
    "tab_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_001887535.1_ASM188753v1_protein',\n",
       " 'GCF_013306235.1_ASM1330623v1_protein',\n",
       " 'GCF_013306335.1_ASM1330633v1_protein',\n",
       " 'GCF_013306435.1_ASM1330643v1_protein',\n",
       " 'GCF_013306935.1_ASM1330693v1_protein',\n",
       " 'GCF_013375735.1_ASM1337573v1_protein']"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_for_og = list(tab_prot.columns) \n",
    "genomes_for_og"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Nucleotide sequences + corresponding positions in genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_013306935.1_ASM1330693v1_protein</th>\n",
       "      <th>GCF_013306235.1_ASM1330623v1_protein</th>\n",
       "      <th>GCF_013306435.1_ASM1330643v1_protein</th>\n",
       "      <th>GCF_001887535.1_ASM188753v1_protein</th>\n",
       "      <th>GCF_013306335.1_ASM1330633v1_protein</th>\n",
       "      <th>GCF_013375735.1_ASM1337573v1_protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP_071507848.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'location': (0, 1, 2,...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'location': (0, 1, 2,...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_013210791.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'location': (1846, 18...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'location': (1846, 18...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_173941386.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'location': (3098, 30...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'location': (3098, 30...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_173941387.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'location': (6305, 63...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'location': (6305, 63...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_086005422.1</th>\n",
       "      <td>{'chr': 'NZ_CP052127.1', 'location': (666300, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052097.1', 'location': (693036, ...</td>\n",
       "      <td>{'chr': 'NZ_CP016555.1', 'location': (1164461,...</td>\n",
       "      <td>{'chr': 'NZ_CP052087.1', 'location': (666300, ...</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1889533,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_016726112.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1945036,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_028861767.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1973916,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_023470131.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1980032,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_043885778.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1985998,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_021156181.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'location': (1991008,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11754 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             GCF_013306935.1_ASM1330693v1_protein  \\\n",
       "WP_071507848.1  {'chr': 'NZ_CP052126.1', 'location': (0, 1, 2,...   \n",
       "WP_013210791.1  {'chr': 'NZ_CP052126.1', 'location': (1846, 18...   \n",
       "WP_173941386.1  {'chr': 'NZ_CP052126.1', 'location': (3098, 30...   \n",
       "WP_173941387.1  {'chr': 'NZ_CP052126.1', 'location': (6305, 63...   \n",
       "WP_086005422.1  {'chr': 'NZ_CP052127.1', 'location': (666300, ...   \n",
       "...                                                           ...   \n",
       "WP_016726112.1                                                NaN   \n",
       "WP_028861767.1                                                NaN   \n",
       "WP_023470131.1                                                NaN   \n",
       "WP_043885778.1                                                NaN   \n",
       "WP_021156181.1                                                NaN   \n",
       "\n",
       "               GCF_013306235.1_ASM1330623v1_protein  \\\n",
       "WP_071507848.1                                  NaN   \n",
       "WP_013210791.1                                  NaN   \n",
       "WP_173941386.1                                  NaN   \n",
       "WP_173941387.1                                  NaN   \n",
       "WP_086005422.1                                  NaN   \n",
       "...                                             ...   \n",
       "WP_016726112.1                                  NaN   \n",
       "WP_028861767.1                                  NaN   \n",
       "WP_023470131.1                                  NaN   \n",
       "WP_043885778.1                                  NaN   \n",
       "WP_021156181.1                                  NaN   \n",
       "\n",
       "                             GCF_013306435.1_ASM1330643v1_protein  \\\n",
       "WP_071507848.1                                                NaN   \n",
       "WP_013210791.1                                                NaN   \n",
       "WP_173941386.1                                                NaN   \n",
       "WP_173941387.1                                                NaN   \n",
       "WP_086005422.1  {'chr': 'NZ_CP052097.1', 'location': (693036, ...   \n",
       "...                                                           ...   \n",
       "WP_016726112.1                                                NaN   \n",
       "WP_028861767.1                                                NaN   \n",
       "WP_023470131.1                                                NaN   \n",
       "WP_043885778.1                                                NaN   \n",
       "WP_021156181.1                                                NaN   \n",
       "\n",
       "                              GCF_001887535.1_ASM188753v1_protein  \\\n",
       "WP_071507848.1                                                NaN   \n",
       "WP_013210791.1                                                NaN   \n",
       "WP_173941386.1                                                NaN   \n",
       "WP_173941387.1                                                NaN   \n",
       "WP_086005422.1  {'chr': 'NZ_CP016555.1', 'location': (1164461,...   \n",
       "...                                                           ...   \n",
       "WP_016726112.1                                                NaN   \n",
       "WP_028861767.1                                                NaN   \n",
       "WP_023470131.1                                                NaN   \n",
       "WP_043885778.1                                                NaN   \n",
       "WP_021156181.1                                                NaN   \n",
       "\n",
       "                             GCF_013306335.1_ASM1330633v1_protein  \\\n",
       "WP_071507848.1  {'chr': 'NZ_CP052086.1', 'location': (0, 1, 2,...   \n",
       "WP_013210791.1  {'chr': 'NZ_CP052086.1', 'location': (1846, 18...   \n",
       "WP_173941386.1  {'chr': 'NZ_CP052086.1', 'location': (3098, 30...   \n",
       "WP_173941387.1  {'chr': 'NZ_CP052086.1', 'location': (6305, 63...   \n",
       "WP_086005422.1  {'chr': 'NZ_CP052087.1', 'location': (666300, ...   \n",
       "...                                                           ...   \n",
       "WP_016726112.1                                                NaN   \n",
       "WP_028861767.1                                                NaN   \n",
       "WP_023470131.1                                                NaN   \n",
       "WP_043885778.1                                                NaN   \n",
       "WP_021156181.1                                                NaN   \n",
       "\n",
       "                             GCF_013375735.1_ASM1337573v1_protein  \n",
       "WP_071507848.1                                                NaN  \n",
       "WP_013210791.1                                                NaN  \n",
       "WP_173941386.1                                                NaN  \n",
       "WP_173941387.1                                                NaN  \n",
       "WP_086005422.1  {'chr': 'NZ_CP056086.1', 'location': (1889533,...  \n",
       "...                                                           ...  \n",
       "WP_016726112.1  {'chr': 'NZ_CP056086.1', 'location': (1945036,...  \n",
       "WP_028861767.1  {'chr': 'NZ_CP056086.1', 'location': (1973916,...  \n",
       "WP_023470131.1  {'chr': 'NZ_CP056086.1', 'location': (1980032,...  \n",
       "WP_043885778.1  {'chr': 'NZ_CP056086.1', 'location': (1985998,...  \n",
       "WP_021156181.1  {'chr': 'NZ_CP056086.1', 'location': (1991008,...  \n",
       "\n",
       "[11754 rows x 6 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prot_db.loc['WP_003261125.1', 'GCF_000215325_1_ASM21532v1_protein']\n",
    "#\"WP_003261125.1\" in prot_db.index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load single proteins alignments and extract gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "gaps2tab = pd.DataFrame(columns=tab_prot.columns, index=tab_prot.index)\n",
    "#gaps2tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCF_001887535.1_ASM188753v1_protein</th>\n",
       "      <th>GCF_013306235.1_ASM1330623v1_protein</th>\n",
       "      <th>GCF_013306335.1_ASM1330633v1_protein</th>\n",
       "      <th>GCF_013306435.1_ASM1330643v1_protein</th>\n",
       "      <th>GCF_013306935.1_ASM1330693v1_protein</th>\n",
       "      <th>GCF_013375735.1_ASM1337573v1_protein</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orthogroup</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OG0000225</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000226</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000227</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000228</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000229</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004089</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004091</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004092</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004093</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004094</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3842 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          GCF_001887535.1_ASM188753v1_protein  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                         GCF_013306235.1_ASM1330623v1_protein  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                         GCF_013306335.1_ASM1330633v1_protein  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                         GCF_013306435.1_ASM1330643v1_protein  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                         GCF_013306935.1_ASM1330693v1_protein  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                         GCF_013375735.1_ASM1337573v1_protein  \n",
       "Orthogroup                                                     \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "...                                                       ...  \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "\n",
       "[3842 rows x 6 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for og in ogs:\n",
    "    pmsa = path_p_aln + '/' +og + '.fa'\n",
    "    for record in SeqIO.parse(pmsa, \"fasta\"):  \n",
    "        #print(record.id)\n",
    "        \n",
    "        header = record.id.split('_WP_')\n",
    "        genome_id = header[0].replace('_1_', '.1_')\n",
    "        protien_id = 'WP_' + header[1]\n",
    "        \n",
    "        # EXACT GAPS MAPPING \n",
    "        gaps_bool = []\n",
    "        for aa in record.seq:\n",
    "            if aa == '-':\n",
    "                gaps_bool.append(1) \n",
    "                gaps_bool.append(1) \n",
    "                gaps_bool.append(1) \n",
    "            else:\n",
    "                gaps_bool.append(0) \n",
    "                gaps_bool.append(0) \n",
    "                gaps_bool.append(0)\n",
    "        \n",
    "        prot_db.loc[protien_id, genome_id]['aln_gaps'] = gaps_bool\n",
    "        gaps2tab.loc[og][genome_id] = gaps_bool\n",
    "gaps2tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome-based solution\n",
    "+ for each genome grab the cols of the core OG table\n",
    "+ concatenate the corresponding nucleotides from the GENOME_PROTEIN_DB\n",
    "\n",
    "## TODO length of nucleotides is larger than length of translation*3 ?!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome GCF_001887535.1_ASM188753v1_protein\n",
      "Exported sequence length 3827778\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3827778\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n",
      "Genome GCF_013306235.1_ASM1330623v1_protein\n",
      "Exported sequence length 3827919\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3827919\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n",
      "Genome GCF_013306335.1_ASM1330633v1_protein\n",
      "Exported sequence length 3824796\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3824796\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n",
      "Genome GCF_013306435.1_ASM1330643v1_protein\n",
      "Exported sequence length 3826515\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3826515\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n",
      "Genome GCF_013306935.1_ASM1330693v1_protein\n",
      "Exported sequence length 3824484\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3824484\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n",
      "Genome GCF_013375735.1_ASM1337573v1_protein\n",
      "Exported sequence length 3827109\n",
      "Exported protein accessions 3825198\n",
      "Exported alignment gaps 3825198\n",
      "Exported original positions 3827109\n",
      "Exported chr labels 0\n",
      "Exported strands 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "g2nucl = {}\n",
    "g2pos = {}\n",
    "g2chr = {}\n",
    "g2strand = {}\n",
    "g2acc = {}\n",
    "g2gaps = {}\n",
    "\n",
    "nucl_fasta = open(\"core_OG_nucl.fasta\", \"w\")\n",
    "\n",
    "for g in genomes_for_og:\n",
    "    print('Genome', g)\n",
    "    \n",
    "    #if g == 'GCF_001586135.1_ASM158613v1_protein' or g == \"GCF_001587135.1_ASM158713v1_protein\":\n",
    "    #    continue\n",
    "    \n",
    "    # get corresponding nucleotide positions \n",
    "    g2pos[g] = []\n",
    "    # get corresponding chr ids \n",
    "    g2chr[g] = []\n",
    "    # get corresponding strands\n",
    "    g2strand[g] = []\n",
    "    # get protein accession for each genome position\n",
    "    g2acc[g.split('.1_')[1][:10]] = []\n",
    "    \n",
    "    g2gaps[g] = []\n",
    "    \n",
    "    # get nucleotide sequences for all the core og proteins per genome \n",
    "    nucl_per_genome = []\n",
    "    \n",
    "    for i in prot_db.loc[tab_prot[g].values, g]: \n",
    "        nucl_per_genome.append(str(i['nucl']))\n",
    "        g2pos[g].extend(i['location'])\n",
    "        #g2chr[g].append((i['chr'], len(i['nucl'])))\n",
    "        prot_len = len(i['nucl']) ######################## <-----\n",
    "        for counter in range(len(i['aln_gaps'])):\n",
    "            if DEBUG:\n",
    "                g2chr[g].append(i['chr'])\n",
    "                g2strand[g].append(i['location'].strand)\n",
    "            g2acc[g.split('.1_')[1][:10]].append(i['acc']) # phylip name format\n",
    "        g2gaps[g].extend(i['aln_gaps'])\n",
    "    \n",
    "    nucl_per_genome = \"\".join(nucl_per_genome)\n",
    "    g2nucl[g] = nucl_per_genome\n",
    "    print('Exported sequence length', len(nucl_per_genome))\n",
    "    print('Exported protein accessions', len(g2acc[g.split('.1_')[1][:10]]))\n",
    "    print('Exported alignment gaps', len(g2gaps[g]))\n",
    "    print('Exported original positions', len(g2pos[g]))\n",
    "    print('Exported chr labels', len(g2chr[g]))\n",
    "    print('Exported strands', len(g2strand[g]))\n",
    "    nucl_fasta.write('>'+g+'\\n')\n",
    "    nucl_fasta.write(nucl_per_genome+'\\n')\n",
    "    print()\n",
    "    \n",
    "nucl_fasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SAVE PROTEIN ACC FOR MAPPING\n",
    "\n",
    "with open(aln2acc, 'wb') as handle:\n",
    "    pickle.dump(g2acc, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# replace ambiguous bases with gaps!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n",
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n",
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n",
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n",
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n",
      "before {'T', 'C', 'G', 'A'}\n",
      "after  {'T', 'C', 'G', 'A'}\n"
     ]
    }
   ],
   "source": [
    "for g in g2nucl:\n",
    "    atcg = {'A', 'C', 'T', 'G'}\n",
    "    ambig = set(g2nucl[g]).difference(atcg)\n",
    "    \n",
    "    print(\"before\", set(g2nucl[g]))\n",
    "    for b in ambig:\n",
    "        g2nucl[g] = g2nucl[g].replace(b, '-')\n",
    "    print(\"after \",set(g2nucl[g]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#str(prot_db.loc[\"WP_097388881.1\",\"GCF_002501565.1_ASM250156v1_protein\"]['nucl'])\n",
    "#for v in prot_db['GCF_001587135.1_ASM158713v1_protein'].values:\n",
    "#    print(v)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_001887535.1_ASM188753v1_protein 3827778\n",
      "GCF_013306235.1_ASM1330623v1_protein 3827919\n",
      "GCF_013306335.1_ASM1330633v1_protein 3824796\n",
      "GCF_013306435.1_ASM1330643v1_protein 3826515\n",
      "GCF_013306935.1_ASM1330693v1_protein 3824484\n",
      "GCF_013375735.1_ASM1337573v1_protein 3827109\n"
     ]
    }
   ],
   "source": [
    "msa_fasta = open(msa_out_fasta, \"w\")\n",
    "for g in g2nucl:\n",
    "    print(g, len(g2nucl[g]))\n",
    "    msa_fasta.write('>'+g.split('_')[2][:10]+'\\n')\n",
    "    \n",
    "    i = 0\n",
    "    for gap in g2gaps[g]:\n",
    "        if gap:\n",
    "            msa_fasta.write('-')\n",
    "        else:\n",
    "            msa_fasta.write(g2nucl[g][i])\n",
    "            i = i + 1\n",
    "    msa_fasta.write('\\n')\n",
    "msa_fasta.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert to Phylip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SeqIO.convert(msa_out_fasta, \"fasta\", msa_out_phylip, \"phylip\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export XMFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_out_xmfa = open(msa_out_xmfa, 'w')\n",
    "\n",
    "for og in ogs:\n",
    "    pmsa = path_p_aln + '/' +og + '.fa'\n",
    "    for record in SeqIO.parse(pmsa, \"fasta\"):\n",
    "        #print(record.id)\n",
    "        g_id = record.id.split('_WP')[0].replace('_1_', '.1_')\n",
    "        p_id = record.id.split('_protein_')[-1]\n",
    "        #print(g_id, p_id)\n",
    "        nns = prot_db.loc[p_id, g_id]['nucl']\n",
    "        #print('fffffff',  nns)\n",
    "        msa_out_xmfa.write('>'+record.id.split('_')[3][:10]+'\\n')\n",
    "        \n",
    "        atcg = {'A', 'T', 'G', 'C','-'}\n",
    "        ambig = set(nns).difference(atcg)\n",
    "        #print(ambig)\n",
    "        for b in ambig:\n",
    "            nns = str(nns).replace(b, '-')\n",
    "        #print('aaaaaa',  nns)\n",
    "        \n",
    "        gaps_v = gaps2tab.loc[og, g_id]\n",
    "        i = 0\n",
    "        for gap in gaps_v:\n",
    "            if gap:\n",
    "                msa_out_xmfa.write('-')\n",
    "            else:\n",
    "                msa_out_xmfa.write(nns[i])\n",
    "                i = i + 1\n",
    "        msa_out_xmfa.write('\\n')\n",
    "    msa_out_xmfa.write('=' + '\\n')\n",
    "msa_out_xmfa.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test the position mapping\n",
    "+ grab sequences from the original gb files using g2pos, g2strand and g2chr\n",
    "+ load exported core OG nucleotide fasta file\n",
    "+ compare the sequences \n",
    "\n",
    "(note: reverse compliment!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DEBUG:\n",
    "    # for the compliment strand mapping\n",
    "    from Bio.Seq import Seq\n",
    "    Seq(\"A\").complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DEBUG:\n",
    "    import os\n",
    "    gbks = '/Users/devseeva/Desktop/work/pan-genome-visualization/data/rso_test/input_GenBank'\n",
    "    ccc=0\n",
    "\n",
    "    test_fasta = {}\n",
    "    for entry in os.scandir(gbks):\n",
    "        #if ccc>0:\n",
    "        #    break\n",
    "        ccc = ccc+1\n",
    "        if (entry.path.endswith(\".gbk\") and entry.is_file()):\n",
    "            genome = os.path.basename(entry).replace('_genomic.gbk', '_protein')\n",
    "            test_fasta[genome] = []\n",
    "            print('Genome:', genome)\n",
    "\n",
    "            # map chr id to the sequence\n",
    "            chr_seq = {}\n",
    "            for record in SeqIO.parse(entry.path, \"gb\"):\n",
    "                print('Chr',record.id, 'with', len(record.features), 'features')\n",
    "                chr_seq[record.id] = record.seq\n",
    "\n",
    "            # get corresponding mapping (from above)\n",
    "            mapping_chr = g2chr[genome]\n",
    "            mapping_pos = g2pos[genome]\n",
    "            mapping_str = g2strand[genome]\n",
    "            assert(len(mapping_chr) == len(mapping_pos))\n",
    "\n",
    "            # iterate through the original positions \n",
    "            for i in range(len(mapping_chr)):\n",
    "                if mapping_str[i] == 1:\n",
    "                    test_fasta[genome].append(chr_seq[mapping_chr[i]][mapping_pos[i]])\n",
    "                else:\n",
    "                    compl = str(Seq(chr_seq[mapping_chr[i]][mapping_pos[i]]).complement())\n",
    "                    test_fasta[genome].append(compl)\n",
    "            test_fasta[genome] = \"\".join(test_fasta[genome])\n",
    "            print(\"Core nucl exported:\", len(test_fasta[genome]))\n",
    "\n",
    "            print()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DEBUG:\n",
    "\n",
    "    for record in SeqIO.parse(\"core_OG_nucl.fasta\", \"fasta\"):\n",
    "        print(record.id, len(record.seq))\n",
    "        print('test length', len(record.seq))\n",
    "        print('Test:', test_fasta[record.id] == str(record.seq))\n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "f = \"/Users/devseeva/Desktop/work/pan-genome-visualization/data/ncbi-rso/metainfo.tsv\"\n",
    "x = pd.read_csv(f, sep='\\t')  \n",
    "phys = []\n",
    "for a in x[\"accession\"]:\n",
    "    phys.append(a.split(\".1_\")[1][:10])\n",
    "x[\"phy_id\"] = phys\n",
    "x.to_csv(\"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/inputs/phy4_test_metadata/phy4_test_cleaned_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
