{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pickle\n",
    "import zlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "DEBUG = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OrthoGroups to MSA \n",
    "+ load table with the proteins - orthogrous distribution \n",
    "+ filter the core orthogroups \n",
    "+ load proteins .pkl database\n",
    "+ get the proteins and the genome positions corresponding to the core OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DEBUG:\n",
    "    ids = 'rso_test' #'korea2' #'rso_test'  #\n",
    "\n",
    "    core_ogs = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_'+ids+'/Species_Tree/Orthogroups_for_concatenated_alignment.txt'\n",
    "    ogs2prot = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_'+ids+'/Orthogroups/Orthogroups.tsv'\n",
    "    prot_db = pd.read_pickle('../'+ids+'_outputs/GENOME_PROTEIN_DB_'+ids+'.pkl')\n",
    "    path_p_aln = '/Users/devseeva/Desktop/work/rso/OrthoFinder/orthoF_msa_'+ids+'/MultipleSequenceAlignments/'\n",
    "\n",
    "    msa_out_fasta = '/Users/devseeva/Desktop/work/sm_workflow/snakefiles/'+ids+'_concat.fasta'\n",
    "    msa_out_phylip = '/Users/devseeva/Desktop/work/sm_workflow/snakefiles/'+ids+'_concat.nexus'\n",
    "    msa_out_xmfa_path = '/Users/devseeva/Desktop/work/sm_workflow/snakefiles/'+ids+'_concat.xmfa'\n",
    "    aln2acc = '/Users/devseeva/Desktop/work/sm_workflow/snakefiles/outputs/MSA_ACC_DB_'+ids+'.pkl'\n",
    "else:\n",
    "    core_ogs = snakemake.input[\"core_ogs\"]\n",
    "    ogs2prot = snakemake.input[\"ogs2prot\"]\n",
    "    prot_db = pd.read_pickle(snakemake.input[\"prot_db\"])\n",
    "    path_p_aln = snakemake.input[\"path2aln\"]\n",
    "\n",
    "    msa_out_fasta = snakemake.output[\"msa_fasta\"]\n",
    "    msa_out_phylip = snakemake.output[\"msa_phylip\"]\n",
    "    msa_out_xmfa_path = snakemake.output[\"msa_xmfa\"]\n",
    "    aln2acc = snakemake.output[\"aln2acc\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GCF_013306935.1_ASM1330693v1': 'FJAT1303.F50',\n",
       " 'GCF_013306235.1_ASM1330623v1': 'FJAT445.F50',\n",
       " 'GCF_013306435.1_ASM1330643v1': 'FJAT15304.F6',\n",
       " 'GCF_001887535.1_ASM188753v1': 'FJAT-1458',\n",
       " 'GCF_013306335.1_ASM1330633v1': 'FJAT15353.F50',\n",
       " 'GCF_013375735.1_ASM1337573v1': 'FJAT91-F8'}"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "strainsNames2fileNames = {}\n",
    "for c in prot_db.columns:\n",
    "    strainsNames2fileNames[c.split('$')[1]] = c.split('$')[0]\n",
    "strainsNames2fileNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FJAT1303.F50</th>\n",
       "      <th>FJAT445.F50</th>\n",
       "      <th>FJAT15304.F6</th>\n",
       "      <th>FJAT-1458</th>\n",
       "      <th>FJAT15353.F50</th>\n",
       "      <th>FJAT91-F8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP_071507848.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cE\\x94\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cE\\x94\\...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_013210791.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cM\\x93\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cM\\x93\\...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_173941386.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cM\\x96\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cM\\x96\\...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_071615401.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9c5\\x92\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052096.1', 'nucl': b'x\\x9c5\\x92\\...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9c5\\x92\\...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_071615399.1</th>\n",
       "      <td>{'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cE\\x92Q...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052096.1', 'nucl': b'x\\x9cE\\x92Q...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cE\\x92Q...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_016726112.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9ce\\x94\\...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_028861767.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'nucl': b\"x\\x9ceS\\xc9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_023470131.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'nucl': b\"x\\x9cMT\\xdb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_043885778.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9cM\\x94\\...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_021156181.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9cM\\x95\\...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8460 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     FJAT1303.F50 FJAT445.F50  \\\n",
       "WP_071507848.1  {'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cE\\x94\\...         NaN   \n",
       "WP_013210791.1  {'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cM\\x93\\...         NaN   \n",
       "WP_173941386.1  {'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cM\\x96\\...         NaN   \n",
       "WP_071615401.1  {'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9c5\\x92\\...         NaN   \n",
       "WP_071615399.1  {'chr': 'NZ_CP052126.1', 'nucl': b'x\\x9cE\\x92Q...         NaN   \n",
       "...                                                           ...         ...   \n",
       "WP_016726112.1                                                NaN         NaN   \n",
       "WP_028861767.1                                                NaN         NaN   \n",
       "WP_023470131.1                                                NaN         NaN   \n",
       "WP_043885778.1                                                NaN         NaN   \n",
       "WP_021156181.1                                                NaN         NaN   \n",
       "\n",
       "                                                     FJAT15304.F6 FJAT-1458  \\\n",
       "WP_071507848.1                                                NaN       NaN   \n",
       "WP_013210791.1                                                NaN       NaN   \n",
       "WP_173941386.1                                                NaN       NaN   \n",
       "WP_071615401.1  {'chr': 'NZ_CP052096.1', 'nucl': b'x\\x9c5\\x92\\...       NaN   \n",
       "WP_071615399.1  {'chr': 'NZ_CP052096.1', 'nucl': b'x\\x9cE\\x92Q...       NaN   \n",
       "...                                                           ...       ...   \n",
       "WP_016726112.1                                                NaN       NaN   \n",
       "WP_028861767.1                                                NaN       NaN   \n",
       "WP_023470131.1                                                NaN       NaN   \n",
       "WP_043885778.1                                                NaN       NaN   \n",
       "WP_021156181.1                                                NaN       NaN   \n",
       "\n",
       "                                                    FJAT15353.F50  \\\n",
       "WP_071507848.1  {'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cE\\x94\\...   \n",
       "WP_013210791.1  {'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cM\\x93\\...   \n",
       "WP_173941386.1  {'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cM\\x96\\...   \n",
       "WP_071615401.1  {'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9c5\\x92\\...   \n",
       "WP_071615399.1  {'chr': 'NZ_CP052086.1', 'nucl': b'x\\x9cE\\x92Q...   \n",
       "...                                                           ...   \n",
       "WP_016726112.1                                                NaN   \n",
       "WP_028861767.1                                                NaN   \n",
       "WP_023470131.1                                                NaN   \n",
       "WP_043885778.1                                                NaN   \n",
       "WP_021156181.1                                                NaN   \n",
       "\n",
       "                                                        FJAT91-F8  \n",
       "WP_071507848.1                                                NaN  \n",
       "WP_013210791.1                                                NaN  \n",
       "WP_173941386.1                                                NaN  \n",
       "WP_071615401.1                                                NaN  \n",
       "WP_071615399.1                                                NaN  \n",
       "...                                                           ...  \n",
       "WP_016726112.1  {'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9ce\\x94\\...  \n",
       "WP_028861767.1  {'chr': 'NZ_CP056086.1', 'nucl': b\"x\\x9ceS\\xc9...  \n",
       "WP_023470131.1  {'chr': 'NZ_CP056086.1', 'nucl': b\"x\\x9cMT\\xdb...  \n",
       "WP_043885778.1  {'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9cM\\x94\\...  \n",
       "WP_021156181.1  {'chr': 'NZ_CP056086.1', 'nucl': b'x\\x9cM\\x95\\...  \n",
       "\n",
       "[8460 rows x 6 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_db = prot_db.rename(columns=lambda x: str(x).split('$')[0])\n",
    "prot_db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get core ogs\n",
    "f = open(core_ogs,\"r\")\n",
    "ogs = []\n",
    "for l in f:\n",
    "    ogs.append(l.replace('\\n',''))\n",
    "f.close()\n",
    "##print(ogs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FJAT-1458</th>\n",
       "      <th>FJAT445.F50</th>\n",
       "      <th>FJAT15353.F50</th>\n",
       "      <th>FJAT15304.F6</th>\n",
       "      <th>FJAT1303.F50</th>\n",
       "      <th>FJAT91-F8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orthogroup</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OG0000225</th>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "      <td>WP_003261911.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000226</th>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "      <td>WP_003261929.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000227</th>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "      <td>WP_003261939.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000228</th>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "      <td>WP_003261941.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000229</th>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "      <td>WP_003262123.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004089</th>\n",
       "      <td>WP_155772886.1</td>\n",
       "      <td>WP_148668905.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_148668915.1</td>\n",
       "      <td>WP_155772886.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004091</th>\n",
       "      <td>WP_155773184.1</td>\n",
       "      <td>WP_173940865.1</td>\n",
       "      <td>WP_173941733.1</td>\n",
       "      <td>WP_119889930.1</td>\n",
       "      <td>WP_173952290.1</td>\n",
       "      <td>WP_086706436.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004092</th>\n",
       "      <td>WP_155773186.1</td>\n",
       "      <td>WP_144061950.1</td>\n",
       "      <td>WP_119417730.1</td>\n",
       "      <td>WP_119889926.1</td>\n",
       "      <td>WP_119417730.1</td>\n",
       "      <td>WP_038938529.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004093</th>\n",
       "      <td>WP_155773188.1</td>\n",
       "      <td>WP_155739038.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155772892.1</td>\n",
       "      <td>WP_155773188.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004094</th>\n",
       "      <td>WP_155773189.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "      <td>WP_058907847.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "      <td>WP_058907847.1</td>\n",
       "      <td>WP_021155598.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3842 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 FJAT-1458     FJAT445.F50   FJAT15353.F50    FJAT15304.F6  \\\n",
       "Orthogroup                                                                   \n",
       "OG0000225   WP_003261911.1  WP_003261911.1  WP_003261911.1  WP_003261911.1   \n",
       "OG0000226   WP_003261929.1  WP_003261929.1  WP_003261929.1  WP_003261929.1   \n",
       "OG0000227   WP_003261939.1  WP_003261939.1  WP_003261939.1  WP_003261939.1   \n",
       "OG0000228   WP_003261941.1  WP_003261941.1  WP_003261941.1  WP_003261941.1   \n",
       "OG0000229   WP_003262123.1  WP_003262123.1  WP_003262123.1  WP_003262123.1   \n",
       "...                    ...             ...             ...             ...   \n",
       "OG0004089   WP_155772886.1  WP_148668905.1  WP_148668915.1  WP_148668915.1   \n",
       "OG0004091   WP_155773184.1  WP_173940865.1  WP_173941733.1  WP_119889930.1   \n",
       "OG0004092   WP_155773186.1  WP_144061950.1  WP_119417730.1  WP_119889926.1   \n",
       "OG0004093   WP_155773188.1  WP_155739038.1  WP_155772892.1  WP_155772892.1   \n",
       "OG0004094   WP_155773189.1  WP_021155598.1  WP_058907847.1  WP_021155598.1   \n",
       "\n",
       "              FJAT1303.F50       FJAT91-F8  \n",
       "Orthogroup                                  \n",
       "OG0000225   WP_003261911.1  WP_003261911.1  \n",
       "OG0000226   WP_003261929.1  WP_003261929.1  \n",
       "OG0000227   WP_003261939.1  WP_003261939.1  \n",
       "OG0000228   WP_003261941.1  WP_003261941.1  \n",
       "OG0000229   WP_003262123.1  WP_003262123.1  \n",
       "...                    ...             ...  \n",
       "OG0004089   WP_148668915.1  WP_155772886.1  \n",
       "OG0004091   WP_173952290.1  WP_086706436.1  \n",
       "OG0004092   WP_119417730.1  WP_038938529.1  \n",
       "OG0004093   WP_155772892.1  WP_155773188.1  \n",
       "OG0004094   WP_058907847.1  WP_021155598.1  \n",
       "\n",
       "[3842 rows x 6 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load roteins ids from the OGs table \n",
    "# keep only the core OGs\n",
    "\n",
    "tab_prot = pd.read_csv(ogs2prot, sep='\\t', index_col=0)  \n",
    "tab_prot = tab_prot.loc[ogs]\n",
    "tab_prot = tab_prot.rename(columns=lambda x: str(x).replace('_protein','')) # ncbi file names correction\n",
    "tab_prot = tab_prot.rename(columns=strainsNames2fileNames)\n",
    "tab_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_for_og = list(tab_prot.columns) \n",
    "\n",
    "#for x in ['pe1', 'pe3', 'pe27', 'pe57', 'pe39']:\n",
    "#    genomes_for_og.remove(x)\n",
    "#genomes_for_og"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get Nucleotide sequences + corresponding positions in genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load single proteins alignments and extract gaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FJAT-1458</th>\n",
       "      <th>FJAT445.F50</th>\n",
       "      <th>FJAT15353.F50</th>\n",
       "      <th>FJAT15304.F6</th>\n",
       "      <th>FJAT1303.F50</th>\n",
       "      <th>FJAT91-F8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Orthogroup</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OG0000225</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000226</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000227</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000228</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0000229</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004089</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004091</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004092</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004093</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OG0004094</th>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3842 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    FJAT-1458  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                  FJAT445.F50  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                FJAT15353.F50  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                 FJAT15304.F6  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                 FJAT1303.F50  \\\n",
       "Orthogroup                                                      \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                       ...   \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                    FJAT91-F8  \n",
       "Orthogroup                                                     \n",
       "OG0000225   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000226   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000227   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000228   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0000229   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "...                                                       ...  \n",
       "OG0004089   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004091   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004092   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004093   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "OG0004094   [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "\n",
       "[3842 rows x 6 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gaps2tab = pd.DataFrame(columns=tab_prot.columns, index=tab_prot.index)\n",
    "\n",
    "len_counter = 0\n",
    "\n",
    "for og in ogs:\n",
    "    pmsa = path_p_aln + '/' +og + '.fa'\n",
    "    genome_ids = []\n",
    "    #if len_counter >= 395140:\n",
    "      #  print(og, '-----')\n",
    "        #len_counter = 0\n",
    "            \n",
    "    for record in SeqIO.parse(pmsa, \"fasta\"):  \n",
    "        #print(pmsa)\n",
    "        \n",
    "        # prokka/ncbi genomes header structure \n",
    "        header = record.id.split('_')\n",
    "        protein_id = header[-2]+'_'+header[-1]\n",
    "        genome_id = strainsNames2fileNames[record.id.replace(protein_id, '').replace('protein_', '')[:-1].replace('_1_', '.1_')]\n",
    "        #genome_ids.append(genome_id)\n",
    "        prot_db.loc[protein_id, genome_id]['og'] = og\n",
    "        \n",
    "        #print(pmsa)\n",
    "        #print(protein_id, genome_id)\n",
    "        \n",
    "        # GAPS PROFILE of aa alignment  \n",
    "        gaps_bool = []\n",
    "        \n",
    "        #if og == 'OG0000732' and genome_id == 'to53':\n",
    "            #print('---', len(record.seq) )\n",
    "        \n",
    "        for aa in record.seq:\n",
    "            if aa == '-':\n",
    "                gaps_bool.append(1) \n",
    "                gaps_bool.append(1) \n",
    "                gaps_bool.append(1) \n",
    "            else:\n",
    "                gaps_bool.append(0) \n",
    "                gaps_bool.append(0) \n",
    "                gaps_bool.append(0)\n",
    "        \n",
    "        \n",
    "        #if genome_id == 'to53':\n",
    "        #    print(gaps_bool)\n",
    "        #    print('to53',record.seq)\n",
    "        #if genome_id == 'to7':\n",
    "        #    print(gaps_bool)\n",
    "        #    print('to7', record.seq)\n",
    "        prot_db.loc[protein_id, genome_id]['aln_gaps'] = gaps_bool\n",
    "        prot_db.loc[protein_id, genome_id]['og'] = og\n",
    "        gaps2tab.loc[og][genome_id] = gaps_bool\n",
    "        \n",
    "        #if len(record.seq)*3 != len(prot_db.loc[protein_id, genome_id]['nucl']):\n",
    "            #print(protein_id, genome_id)\n",
    "            #print('ffffff',len(record.seq)*3, len(prot_db.loc[protein_id, genome_id]['nucl']))\n",
    "    #print(og, genome_ids)\n",
    "gaps2tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import ast\n",
    "d = ast.literal_eval(prot_db.loc[protein_id, genome_id])\n",
    "d\n",
    "d['og'] = og\n",
    "prot_db.loc[protein_id, genome_id] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "ogs_to_drop_from_core = []\n",
    "for g in genomes_for_og:\n",
    "    for og in ogs:\n",
    "        \n",
    "        msa_l = len(prot_db.loc[tab_prot.loc[og, g], g]['aln_gaps'])\n",
    "        seq_l = len(zlib.decompress(prot_db.loc[tab_prot.loc[og, g], g]['nucl']).decode())\n",
    "        gap_l = gaps2tab.loc[og, g].count(1)\n",
    "        \n",
    "        if seq_l + gap_l != msa_l:\n",
    "            print('Length error in:', g, og, ': MSA length:', msa_l, 'protein length:' ,seq_l)\n",
    "            ogs_to_drop_from_core.append(og)\n",
    "            \n",
    "tab_prot = tab_prot.drop(ogs_to_drop_from_core)\n",
    "gaps2tab = gaps2tab.drop(ogs_to_drop_from_core)\n",
    "ogs = [i for i in ogs if i not in ogs_to_drop_from_core]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'\\x1f\\x8b\\x08\\x00\\x0f|Eb\\x02\\xff\\xf3\\xc9/)V\\xc8OSH\\xce\\xcf+I\\xcd+Q\\xc8H-J\\x05\\x00~u \\x8e\\x14\\x00\\x00\\x00'"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import gzip\n",
    "s_in = b\"Lots of content here\"\n",
    "s_out = gzip.compress(s_in)\n",
    "s_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'x\\x9cUU\\xd1\\x95\\xc30\\x08\\x9bM\\x8f\\x0f\\x16\\xd0\\xfe\\xb3\\x1cH\\xc2\\xe9\\xb5i\\x9b\\xd8\\x18\\x84\\x10\\x14lT7\\xbaP\\x98\\xefj\\xbd\\xd1l\\xd2\\x0f\\xb3\\x8bn\\xdb`\\x96\\xc7h~\\xc6d\\xb6lRZ\\xdc\\xbby\\x9e\\xaf\\xb1\\x9d\\xcf\\\\\\xbbX\\xd4G\\xd6@\\xac[A\\xb8\\x8b\\x8a\\xb6\\xab\\xe7d.\\x99e\\x8f\\xf2Sq8[\\x95X\\xb21f?\\xb6P\\xb5\\x8f\\xb2}\\xa4\\xb2%`\\x86\\xa4\\xe8p$\\xa7/L\\x1co\\xba\\x0eAB8\\xa5\\xf1<x!HuG\\x0c\\x06\\xa2\\xae\\xef\\xbe\\xcfsQ\\xfcP\\x80X\\xc6\\xef\\xb3\\xb8\\x9b\\xcb\\xc59(G\\x88y\\xca&\\xe0\\x06\\xb9\\xec\\x10\\xa6g\\x97\\xa9\\x97L\\x94\\xcf&\\xe0D\\xa9d\\xb8\\x80\\x8d!\\xc4\\x89\\x9c\\x8d\\xe0\\xf4\\xb5\\xe6\\n\\x82)\\x04^\\xb1V\\x10\\xaeS\\xd7\\xab\\xa6v\\xb9\\x07\\xc3SE\\'\\xa5\\x83\\x82Q\\xf6\\x9b\\xa0Tz\\x8b_\\xebA\\xb2 \\xe6\\xb5\\xebx\\t2J\\xa0\\xf0-\\xa5N\\xb6we\\xaeT\\x9c\\x16C\\\\-\\xe1b\\xff\\x83j\\x0fR)\\x92^\\xd8\\x95`8\\x8b\\xe4\\x13\\xdb\\x9a\\xd1\\xd2\\x15\\xf3\\xfc\\xd5\\x0f\\xda\\t\\xa45\\xa4~\\xe0\\x08w\\x95G\\xa1\\x97\\x85\\x99\\x84v\\xbd\\x82J\\ra\\xc6\\x19W\\x93|\\xc0[\\x90LD\\xe3\\x92B,\\xbcT>]\\xc3p\\xbe\\r*f \\xe4i+\\xb2O\\xb6\\xb4\\x97\\xc3\\xdca@\\xa9\\xb8\\xd0*\\n\\x94\\xf5RH\\x1fa\\xba\\xbf~b=\\x95\"\\xe5\\x88\\x8c=\\x03\\xeax\\xb8\\x1et\\xcd\\xd7\\x1b\\x8c\\xaa^}\"\\x15\\x083\\xce\\xe7\\xa3\\xd7\\xed\\xf6\\xdd\\xe4\\xae\\xb7bm\\xd9X\\xe2\\x82\\x8d\\x9b0\\x06\\xa6\\xf9\"\\r\\xca\\x18\"\\x93\\x19N\\xff\\x1b\\xc1h\\xa4P\\xcb\\xcbBj5\\x917\\xa3%\\xc5\\x81\\xc9\\xbc\\xd6;\\xedi\\xcc\\xa5\\xc1W\\xa2\\x94x<\\xbb<&R\\x00\\xdeXsk@F\\x9e\\x8c\\xe6{\\x0f\\xe2F\\xa8u\\'\\xfd\\x9f\\xb4\\xcb\\xf8\\xe1!\\xe6\\xa0\\xd7\\x9b\\x8d4{\\xbd\\x82\\x8a\\xdb\\x9b\\xbc7\\x84\\xed[s>S;3Yc\\x8d\\xdd\\xc7H\\x84\\xbaP7#7E\\t64\\xdf\\xacC\\xdaYD\\xdd\\x9e\\x0c\\x06\\x99\\xb8\\xc8@\\x06\\xa2\"\\x9cY\\xabM\\xd4:j\\xd5\\xbe\\xe1\\xe6\\xae\\x87\\x02\\xe2\\x1b\\x9c\\xfeGQ\\xd7(q\\xff\\xfb\\xa8\\x10\\xcfy[\\x88\\xd8f\\x10l\\xcf\\x07f\\x14 \\xf9\\x96M\\x98\\x11\\xf4\\xba\\xe2\\xb4\\x90N\\xf4\\xe2\\x1f\\x1d#\\xf6\\xaf'"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_db.loc[tab_prot.loc[og, g], g]['nucl']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genome-based solution\n",
    "+ for each genome grab the cols of the core OG table\n",
    "+ concatenate the corresponding nucleotides from the GENOME_PROTEIN_DB\n",
    "\n",
    "## TODO length of nucleotides is larger than length of translation*3 ?!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome FJAT-1458\n",
      "Exported sequence length NO GAPS 3816252\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n",
      "Genome FJAT445.F50\n",
      "Exported sequence length NO GAPS 3816393\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n",
      "Genome FJAT15353.F50\n",
      "Exported sequence length NO GAPS 3813270\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n",
      "Genome FJAT15304.F6\n",
      "Exported sequence length NO GAPS 3814989\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n",
      "Genome FJAT1303.F50\n",
      "Exported sequence length NO GAPS 3812958\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n",
      "Genome FJAT91-F8\n",
      "Exported sequence length NO GAPS 3815583\n",
      "Exported alignment WITH GAPS 3825198\n",
      "Exported protein accessions 3825198\n",
      "Exported numeric positions 0\n",
      "Exported chromosomal positions 0\n",
      "Exported strands orientations 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "g2nucl = {}\n",
    "g2pos = {}\n",
    "g2chr = {}\n",
    "g2strand = {}\n",
    "g2acc = {}\n",
    "g2gaps = {}\n",
    "g2og = {}\n",
    "\n",
    "for g in genomes_for_og:\n",
    "    print('Genome', g)\n",
    "    \n",
    "    # get corresponding nucleotide positions \n",
    "    g2pos[g] = []\n",
    "    # get corresponding chr ids \n",
    "    g2chr[g] = []\n",
    "    # get corresponding strands\n",
    "    g2strand[g] = []\n",
    "    # get protein accession for each genome position\n",
    "    g2acc[g] = []\n",
    "    # get gaps profile\n",
    "    g2gaps[g] = []\n",
    "    # debug\n",
    "    g2og[g] = []\n",
    "    \n",
    "    # get nucleotide sequences for all the core og proteins per genome \n",
    "    nucl_per_genome = []\n",
    "    #print(prot_db.loc[tab_prot[g].values, g])\n",
    "    \n",
    "    for i in prot_db.loc[tab_prot[g].values, g]: \n",
    "        \n",
    "        nucl_per_genome.append(zlib.decompress(i['nucl']).decode())\n",
    "        #g2pos[g].extend(i['location'])\n",
    "        #print('----',g, )\n",
    "        g2og[g].append(i['og'])\n",
    "        \n",
    "        for counter in range(len(i['aln_gaps'])):\n",
    "            #g2chr[g].append(i['chr'])\n",
    "            #g2strand[g].append(i['location'].strand)\n",
    "            g2acc[g].append(i['acc'])\n",
    "        g2gaps[g].extend(i['aln_gaps'])\n",
    "    \n",
    "    nucl_per_genome = \"\".join(nucl_per_genome)\n",
    "    g2nucl[g] = nucl_per_genome\n",
    "    print('Exported sequence length NO GAPS', len(nucl_per_genome))\n",
    "    print('Exported alignment WITH GAPS', len(g2gaps[g]))\n",
    "    print('Exported protein accessions', len(g2acc[g]))\n",
    "    print('Exported numeric positions', len(g2pos[g]))\n",
    "    print('Exported chromosomal positions', len(g2chr[g]))\n",
    "    print('Exported strands orientations', len(g2strand[g]))\n",
    "    print()\n",
    "    \n",
    "    ####### debug \n",
    "    #g2nucl[g] = g2nucl[g][:395150]\n",
    "    #g2gaps[g] = g2gaps[g][:395150]\n",
    "    #g2acc[g] = g2acc[g][:395150]\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MSA repeats problem \n",
    "\n",
    "some proteins sequence of the OrthoFinder may have long repeat sequences, which are not present in the orthologs. In the resulting MSA produced by OrthoFinder the repeated sequences are skipped:\n",
    "\n",
    "e.g. \n",
    "+ original input: <br>\n",
    "\\> p1 <br>\n",
    "ABCXYZ <br>\n",
    "\\> p2 <br>\n",
    "ABCABCXYZ <br>\n",
    "+ MSA output:<br>\n",
    "\\> p1 <br>\n",
    "ABCXYZ <br>\n",
    "\\> p2 <br>\n",
    "ABCXYZ <br>\n",
    "\n",
    "This results in large shifts in the final nucleotides MSA <br>\n",
    "\n",
    "Examples from to53\n",
    "\n",
    "OG: OG0000732; gene: DEOEPAP_01560\n",
    "MNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHTVNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHTVVIGDDGRALEVQIRTREMHHFAEYGVAAHWRYKEAGSKGYAGQFSASERYDEKIAWLRQLLAWKDDAEHTVAQEDSPWEQLKHTELDDHIYVLTPQARVIALPQGATPVDFAYYLHSDLGHRCRGARVDGTMVPLNTPLKNGQTVEIVTVKQGGPSRDWLNPELHYLASNRARAKVRAWFNALELEETLAQGRILIDKTLQREGKTAVNLEELAAKLGFKTPDELYAMVAKDEFSLRHVEAVLRTDGAPPAPADDEASITKKSRATSVARGAKSGVLVVGVDSLLTQMSRCCKPAPPDLIVGFVTRGRGVSIHRQNCATFQQLAARAPGRVIQTEWGNRGEAVYPVDIHVEALDRQGLLRDISEILSREKINVTGVRTLSSKGVAKMQFTAEVSEATQLQRALILIEEVTGVLSARRK\n",
    "VNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHT                                                                                                                                                                                                                                                                                                                                       VVIGDDGRALEVQIRTREMHHFAEYGVAAHWRYKEAGSKGYAGQFSASERYDEKIAWLRQLLAWKDDAEHTVAQEDSPWEQLKHTELDDHIYVLTPQARVIALPQGATPVDFAYYLHSDLGHRCRGARVDGTMVPLNTPLKNGQTVEIVTVKQGGPSRDWLNPELHYLASNRARAKVRAWFNALELEETLAQGRILIDKTLQREGKTAVNLEELAAKLGFKTPDELYAMVAKDEFSLRHVEAVLRTDGAPPAPADDEASITKKSRATSVARGAKSGVLVVGVDSLLTQMSRCCKPAPPDLIVGFVTRGRGVSIHRQNCATFQQLAARAPGRVIQTEWGNRGEAVYPVDIHVEALDRQGLLRDISEILSREKINVTGVRTLSSKGVAKMQFTAEVSEATQLQRALILIEEVTGVLSARRK\n",
    "\n",
    "OG: OG0002212; gene: GDEOEPAP_02088\n",
    "MTRVDLRSAKPDYDGPLLRLDDLSVRFEAEHNEWIEAVRHVSFDLHAGERFALVGESGSGKTVTALSIMRLLADAHYSGRILFEGRDLLAASERQMRGLRGADVAMVFQEPMTALNPLYTIGNQIVETLELHEGLDKRAAKARAIALLEHTGIAEAPRRFDAFPHQLSGGQRQRAMIAMALACSPKLLVADEPTTALDVTVRMQILQLLRDLQAEYGMAVMLITHDLNMVRAFAERVGVMERGVLVETGETATVFSAPQHPYTVRLLESRPQRDVLPLVPLAPVLLEADQLTVTYDRHRPGFAGWFRADPFTAVDQVSLQLREGETLGIVGESGSGKTTLAQTLLGLQRPKAGDLRFLGSSLLGQTRDERRAVRARMQVVFQDPFGSLSPRMTIEQIVGEGLALHQPHVNDAERRQRVIEALREVGLDRTALGRYPHEFSGGQRQRIAIARVLILKPQVLVLDEPTSALDVSIQQQVLSLLSALQKKYNLSYLFISHDLAVIRAMSHRVMVMKDGKVVEAGDTEAVLASPSHPYTCKLMAAAALKPLQPLQPPG\n",
    "MTRVDLRSAKPDYDGPLLRLDDLSVRFEAEHNEWIEAVRHVSFDLHAGERFALVGESGSGKTVTALSIMRLLADAHYSGRILFEGRDLLAASERQMRGLRGADVAMVFQEPMTALNPLYTIGNQIVETLELHEGLDKRAAKARAIALLEHTGIAEAPRRFDAFPHQLSGGQRQRAMIAMALACSPKLLVADEPTTALDVTVRMQILQLLRDLQAEYGMAVMLITHDLNMVRAFAERVGVMERGVLVETGETATVFSAPQHPYTVRLLESRPQRDVLPLVPLAPVLLEADQLTVTYDRHRPGFAGWFRADPFTAVDQVSLQLREGETLGIVGESGSGKTTLAQTLLGLQRPKAGDLRFLGSSLLGQTRDERRAVRARMQVVFQDPFGSLSPRMTIEQIVGEGLALHQPHVNDAERRQRVIEALREVGLDRTALGRYPHEFSGGQRQRIAIARVLILKPQVLVLDEPTSALDVSIQQQVLSLLSALQKKYNLSYLFISHDLAVIRAMSHRVMVMKDGKVVEAGDTEAVLASPSHPYTCKLMAAAALKPLQ   PPG\n",
    "\n",
    "OG: OG0003692; gene: GDEOEPAP_04673\n",
    "MRHFPTLHSVALIPSGAAITRAAVTGHARHPRAPVRNALGGTVLLGTFAAAIPAFAAPAGGTVTSGSGSIGQNGTTTTITQTSSRLAIDWSAFGIQAGETVNFIQPGAGAIALNRVTGHEATSIFGSLNANGTVFILNPNGVLFGGGAQVNVGGLVASTLGLSNADFEAGRYALSGGSTAGVTNQGTITVPSGGKVALIANSVDNAGNISAPRGSVLLAGAGNVTLTLADGSPLGYTISAGAARTLVNNGGMIVADGGRVVLTARGLDSLSESVVNTTGVVRARTVGNNQGTIELIGDPMAGLTQVSGQIDASAPDGGGNGGSVKVLGTKVGVFDGARIDVSGMAGGGTALIGGNAQGAGPEPNATATYVAPTAAIDASAIRQGNGGRLIVWGTDVANVHGSLSANGGAQGGNGGHVETSGHALDTDGIAVSVAGGGGG  TGGLWLLDPYNVTLSAGTQTGGAFSGNVWTPSASGSLVNTSSIQSLLNSGGNVTITTTGAGTQEGNIAINGSISKTAGGNASLTLIADGRITTNASSGTHRTITSTSGTLDVSMTARATTSAASTSGINLRYVDINANGGNISATASGAQSGTAAALSLQNSSWSTTGAGNISLSGTLPDNGNSQGVFLNANTLTTASGSITVSGTSGGIATITGTNSNGAPLFSTSNIGVYLAGGNTLRSTGGGAIALTGTATGATATWAGSGVTLSALDSLSTSGAVTISGTASNPVSSTYRNQLSTVNIAGTGSNAASLTGGTVTITGTNSVVGNASSTSNGNAAVKLDGKVNLTATAGPINIAGSNAGGDGVWGSGSGAVTMSAPASSSITISARSLDSVSGYSGFYIGGGSATLTFATAAPVSIAAESLVGARKAFWNKGGLTVPGNLSIVTTGGAIADDTTHGGYFHVTGSTSIDSSGAGNTVSLTNSGNTFAGAISLNAGDTTLASSTSLTLDSSTVNGALNLTAPGITQSGAVSAAGTATLNGGNAAITLTNDGNAFNALAVTGAAGVNVVNANALSVTGVTSTGAVRFTTLSGDLATRGNWSISGGDLTLSAGAGNTRGTASGGNVDSQATLAVDAGRTITVYSGAVGSTVLGGTLAGRAAAGSGNFRYNRQDGDAPGADGIGDGATYVMYRERPTVTVTPTDAANTKVYDGGGASDPALAYTIVGQVNGDSATQILSGSLARAAGQNAGRYAISQGTLADQLGYQTVLDVGHAFSITPAPLLITANGASRTAGQDNPAFSATFAGFVGGQTAQTADLRGTLRLLTSADAGSPAGSYAIVPDGISSNNYSIRYAAGTLSVLPASPAPMVAVPGEPSYMNAVIDMQARIRAAGLASAPEPAERSATGWRTTTATFRQDDTASDNVPGNRLAAQE\n",
    "MRHFPTLHSVALIPSGAAITRAAVTGHARHPRAPVRNALGGTVLLGTFAAAIPAFAAPAGGTVTSGSGSIGQNGTTTTITQTSSRLAIDWSAFGIQAGETVNFIQPGAGAIALNRVTGHEATSIFGSLNANGTVFILNPNGVLFGGGAQVNVGGLVASTLGLSNADFEAGRYALSGGSTAGVTNQGTITVPSGGKVALIANSVDNAGNISAPRGSVLLAGAGNVTLTLADGSPLGYTISAGAARTLVNNGGMIVADGGRVVLTARGLDSLSESVVNTTGVVRARTVGNNQGTIELIGDPMAGLTQVSGQIDASAPDGGGNGGSVKVLGTKVGVFDGARIDVSGMAGGGTALIGGNAQGAGPEPNATATYVAPTAAIDASAIRQGNGGRLIVWGTDVANVHGSLSANGGAQGGNGGHVETSGHALDTDGIAVSVAGGGGGGGTGGLWLLDPYNVTLSAGTQTGGAFSGNVWTPSASGSLVNTSSIQSLLNSGGNVTITTTGAGTQEGNIAINGSISKTAGGNASLTLIADGRITTNASSGTHRTITSTSGTLDVSMTARATTSAASTSGINLRYVDINANGGNISATASGAQSGTAAALSLQNSSWSTTGAGNISLSGTLPDNGNSQGVFLNANTLTTASGSITVSGTSGGIATITGTNSNGAPLFSTSNIGVYLAGGNTLRSTGGGAIALTGTATGATATWAGSGVTLSALDSLSTSGAVTISGTASNPVSSTYRNQLSTVNIAGTGSNAASLTGGTVTITGTNSVVGNASSTSNGNAAVKLDGKVNLTATAGPINIAGSNAGGDGVWGSGSGAVTMSAPASSSITISARSLDSVSGYSGFYIGGGSATLTFATAAPVSIAAESLVGARKAFWNKGGLTVPGNLSIVTTGGAIADDTTHGGYFHVTGSTSIDSSGAGNTVSLTNSGNTFAGAISLNAGDTTLASSTSLTLDSSTVNGALNLTAPGITQSGAVSAAGTATLNGGNAAITLTNDGNAFNALAVTGAAGVNVVNANALSVTGVTSTGAVRFTTLSGDLATRGNWSISGGDLTLSAGAGNTRGTASGGNVDSQATLAVDAGRTITVYSGAVGSTVLGGTLAGRAAAGSGNFRYNRQDGDAPGADGIGDGATYVMYRERPTVTVTPTDAANTKVYDGGGASDPALAYTIVGQVNGDSATQILSGSLARAAGQNAGRYAISQGTLADQLGYQTVLDVGHAFSITPAPLLITANGASRTAGQDNPAFSATFAGFVGGQTAQTADLRGTLRLLTSADAGSPAGSYAIVPDGISSNNYSIRYAAGTLSVLPASPAPMVAVPGEPSYMNAVIDMQARIRAAGLASAPEPAERSATGWRTTTATFRQDDTASDNVPGNRLAAQE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['- M', '- N', '- S', '- K', '- T', '- D', '- R', '- P', '- A', '- E', '- G', '- A', '- G', '- A', '- T', '- A', '- D', '- R', '- A', '- A', '- R', '- E', '- A', '- A', '- V', '- A', '- R', '- A', '- L', '- E', '- C', '- L', '- A', '- D', '- H', '- A', '- G', '- D', '- H', '- Q', '- T', '- L', '- T', '- G', '- E', '- P', '- L', '- V', '- S', '- H', '- A', '- R', '- G', '- T', '- V', '- A', '- I', '- L', '- E', '- G', '- L', '- R', '- V', '- D', '- A', '- P', '- S', '- L', '- Q', '- A', '- A', '- A', '- L', '- F', '- L', '- L', '- P', '- I', '- L', '- A', '- T', '- D', '- N', '- E', '- K', '- A', '- I', '- E', '- P', '- A', '- F', '- G', '- G', '- E', '- V', '- A', '- R', '- L', '- V', '- H', '- D', '- V', '- R', '- Q', '- L', '- L', '- R', '- I', '- G', '- A', '- I', '- A', '- G', '- M', '- A', '- S', '- P', '- A', '- E', '- A', '- G', '- V', '- T', '- R', '- K', '- N', '- E', '- A', '- E', '- A', '- R', '- R', '- A', '- Q', '- V', '- E', '- A', '- L', '- R', '- K', '- M', '- L', '- L', '- A', '- F', '- A', '- Q', '- D', '- I', '- R', '- V', '- V', '- L', '- I', '- R', '- L', '- A', '- S', '- R', '- L', '- Q', '- S', '- L', '- R', '- W', '- L', '- A', '- Q', '- N', '- K', '- L', '- P', '- A', '- P', '- E', '- G', '- M', '- A', '- R', '- E', '- T', '- L', '- D', '- I', '- Y', '- A', '- P', '- L', '- A', '- N', '- R', '- L', '- G', '- I', '- W', '- Q', '- L', '- K', '- W', '- E', '- L', '- E', '- D', '- L', '- G', '- F', '- R', '- F', '- E', '- D', '- P', '- D', '- T', '- Y', '- K', '- R', '- I', '- A', '- R', '- L', '- L', '- D', '- E', '- K', '- R', '- I', '- E', '- R', '- E', '- Q', '- F', '- I', '- G', '- E', '- A', '- I', '- A', '- R', '- L', '- Q', '- Q', '- T', '- L', '- R', '- D', '- A', '- G', '- I', '- H', '- A', '- E', '- V', '- S', '- G', '- R', '- P', '- K', '- H', '- I', '- Y', '- S', '- I', '- W', '- R', '- K', '- M', '- R', '- G', '- K', '- D', '- L', '- D', '- F', '- A', '- D', '- L', '- Y', '- D', '- V', '- R', '- A', '- F', '- R', '- V', '- I', '- V', '- D', '- D', '- I', '- K', '- D', '- C', '- Y', '- T', '- V', '- L', '- G', '- F', '- V', '- H', '- H', '- M', '- W', '- Q', '- P', '- I', '- P', '- K', '- E', '- F', '- D', '- D', '- Y', '- I', '- S', '- R', '- P', '- K', '- T', '- N', '- G', '- Y', '- K', '- S', '- L', '- H', '- T']\n",
      "rep: MNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHT\n",
      "2\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "import difflib\n",
    "\n",
    "s_nucl =\"MTRVDLRSAKPDYDGPLLRLDDLSVRFEAEHNEWIEAVRHVSFDLHAGERFALVGESGSGKTVTALSIMRLLADAHYSGRILFEGRDLLAASERQMRGLRGADVAMVFQEPMTALNPLYTIGNQIVETLELHEGLDKRAAKARAIALLEHTGIAEAPRRFDAFPHQLSGGQRQRAMIAMALACSPKLLVADEPTTALDVTVRMQILQLLRDLQAEYGMAVMLITHDLNMVRAFAERVGVMERGVLVETGETATVFSAPQHPYTVRLLESRPQRDVLPLVPLAPVLLEADQLTVTYDRHRPGFAGWFRADPFTAVDQVSLQLREGETLGIVGESGSGKTTLAQTLLGLQRPKAGDLRFLGSSLLGQTRDERRAVRARMQVVFQDPFGSLSPRMTIEQIVGEGLALHQPHVNDAERRQRVIEALREVGLDRTALGRYPHEFSGGQRQRIAIARVLILKPQVLVLDEPTSALDVSIQQQVLSLLSALQKKYNLSYLFISHDLAVIRAMSHRVMVMKDGKVVEAGDTEAVLASPSHPYTCKLMAAAALKPLQPLQPPG\"\n",
    "s_msa = \"MTRVDLRSAKPDYDGPLLRLDDLSVRFEAEHNEWIEAVRHVSFDLHAGERFALVGESGSGKTVTALSIMRLLADAHYSGRILFEGRDLLAASERQMRGLRGADVAMVFQEPMTALNPLYTIGNQIVETLELHEGLDKRAAKARAIALLEHTGIAEAPRRFDAFPHQLSGGQRQRAMIAMALACSPKLLVADEPTTALDVTVRMQILQLLRDLQAEYGMAVMLITHDLNMVRAFAERVGVMERGVLVETGETATVFSAPQHPYTVRLLESRPQRDVLPLVPLAPVLLEADQLTVTYDRHRPGFAGWFRADPFTAVDQVSLQLREGETLGIVGESGSGKTTLAQTLLGLQRPKAGDLRFLGSSLLGQTRDERRAVRARMQVVFQDPFGSLSPRMTIEQIVGEGLALHQPHVNDAERRQRVIEALREVGLDRTALGRYPHEFSGGQRQRIAIARVLILKPQVLVLDEPTSALDVSIQQQVLSLLSALQKKYNLSYLFISHDLAVIRAMSHRVMVMKDGKVVEAGDTEAVLASPSHPYTCKLMAAAALKPLQPPG\"\n",
    "\n",
    "s_msa = \"MRHFPTLHSVALIPSGAAITRAAVTGHARHPRAPVRNALGGTVLLGTFAAAIPAFAAPAGGTVTSGSGSIGQNGTTTTITQTSSRLAIDWSAFGIQAGETVNFIQPGAGAIALNRVTGHEATSIFGSLNANGTVFILNPNGVLFGGGAQVNVGGLVASTLGLSNADFEAGRYALSGGSTAGVTNQGTITVPSGGKVALIANSVDNAGNISAPRGSVLLAGAGNVTLTLADGSPLGYTISAGAARTLVNNGGMIVADGGRVVLTARGLDSLSESVVNTTGVVRARTVGNNQGTIELIGDPMAGLTQVSGQIDASAPDGGGNGGSVKVLGTKVGVFDGARIDVSGMAGGGTALIGGNAQGAGPEPNATATYVAPTAAIDASAIRQGNGGRLIVWGTDVANVHGSLSANGGAQGGNGGHVETSGHALDTDGIAVSVAGGGGGTGGLWLLDPYNVTLSAGTQTGGAFSGNVWTPSASGSLVNTSSIQSLLNSGGNVTITTTGAGTQEGNIAINGSISKTAGGNASLTLIADGRITTNASSGTHRTITSTSGTLDVSMTARATTSAASTSGINLRYVDINANGGNISATASGAQSGTAAALSLQNSSWSTTGAGNISLSGTLPDNGNSQGVFLNANTLTTASGSITVSGTSGGIATITGTNSNGAPLFSTSNIGVYLAGGNTLRSTGGGAIALTGTATGATATWAGSGVTLSALDSLSTSGAVTISGTASNPVSSTYRNQLSTVNIAGTGSNAASLTGGTVTITGTNSVVGNASSTSNGNAAVKLDGKVNLTATAGPINIAGSNAGGDGVWGSGSGAVTMSAPASSSITISARSLDSVSGYSGFYIGGGSATLTFATAAPVSIAAESLVGARKAFWNKGGLTVPGNLSIVTTGGAIADDTTHGGYFHVTGSTSIDSSGAGNTVSLTNSGNTFAGAISLNAGDTTLASSTSLTLDSSTVNGALNLTAPGITQSGAVSAAGTATLNGGNAAITLTNDGNAFNALAVTGAAGVNVVNANALSVTGVTSTGAVRFTTLSGDLATRGNWSISGGDLTLSAGAGNTRGTASGGNVDSQATLAVDAGRTITVYSGAVGSTVLGGTLAGRAAAGSGNFRYNRQDGDAPGADGIGDGATYVMYRERPTVTVTPTDAANTKVYDGGGASDPALAYTIVGQVNGDSATQILSGSLARAAGQNAGRYAISQGTLADQLGYQTVLDVGHAFSITPAPLLITANGASRTAGQDNPAFSATFAGFVGGQTAQTADLRGTLRLLTSADAGSPAGSYAIVPDGISSNNYSIRYAAGTLSVLPASPAPMVAVPGEPSYMNAVIDMQARIRAAGLASAPEPAERSATGWRTTTATFRQDDTASDNVPGNRLAAQE\"\n",
    "s_nucl =  \"MRHFPTLHSVALIPSGAAITRAAVTGHARHPRAPVRNALGGTVLLGTFAAAIPAFAAPAGGTVTSGSGSIGQNGTTTTITQTSSRLAIDWSAFGIQAGETVNFIQPGAGAIALNRVTGHEATSIFGSLNANGTVFILNPNGVLFGGGAQVNVGGLVASTLGLSNADFEAGRYALSGGSTAGVTNQGTITVPSGGKVALIANSVDNAGNISAPRGSVLLAGAGNVTLTLADGSPLGYTISAGAARTLVNNGGMIVADGGRVVLTARGLDSLSESVVNTTGVVRARTVGNNQGTIELIGDPMAGLTQVSGQIDASAPDGGGNGGSVKVLGTKVGVFDGARIDVSGMAGGGTALIGGNAQGAGPEPNATATYVAPTAAIDASAIRQGNGGRLIVWGTDVANVHGSLSANGGAQGGNGGHVETSGHALDTDGIAVSVAGGGGGGGTGGLWLLDPYNVTLSAGTQTGGAFSGNVWTPSASGSLVNTSSIQSLLNSGGNVTITTTGAGTQEGNIAINGSISKTAGGNASLTLIADGRITTNASSGTHRTITSTSGTLDVSMTARATTSAASTSGINLRYVDINANGGNISATASGAQSGTAAALSLQNSSWSTTGAGNISLSGTLPDNGNSQGVFLNANTLTTASGSITVSGTSGGIATITGTNSNGAPLFSTSNIGVYLAGGNTLRSTGGGAIALTGTATGATATWAGSGVTLSALDSLSTSGAVTISGTASNPVSSTYRNQLSTVNIAGTGSNAASLTGGTVTITGTNSVVGNASSTSNGNAAVKLDGKVNLTATAGPINIAGSNAGGDGVWGSGSGAVTMSAPASSSITISARSLDSVSGYSGFYIGGGSATLTFATAAPVSIAAESLVGARKAFWNKGGLTVPGNLSIVTTGGAIADDTTHGGYFHVTGSTSIDSSGAGNTVSLTNSGNTFAGAISLNAGDTTLASSTSLTLDSSTVNGALNLTAPGITQSGAVSAAGTATLNGGNAAITLTNDGNAFNALAVTGAAGVNVVNANALSVTGVTSTGAVRFTTLSGDLATRGNWSISGGDLTLSAGAGNTRGTASGGNVDSQATLAVDAGRTITVYSGAVGSTVLGGTLAGRAAAGSGNFRYNRQDGDAPGADGIGDGATYVMYRERPTVTVTPTDAANTKVYDGGGASDPALAYTIVGQVNGDSATQILSGSLARAAGQNAGRYAISQGTLADQLGYQTVLDVGHAFSITPAPLLITANGASRTAGQDNPAFSATFAGFVGGQTAQTADLRGTLRLLTSADAGSPAGSYAIVPDGISSNNYSIRYAAGTLSVLPASPAPMVAVPGEPSYMNAVIDMQARIRAAGLASAPEPAERSATGWRTTTATFRQDDTASDNVPGNRLAAQE\"\n",
    "\n",
    "s_nucl =\"MNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHTVNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHTVVIGDDGRALEVQIRTREMHHFAEYGVAAHWRYKEAGSKGYAGQFSASERYDEKIAWLRQLLAWKDDAEHTVAQEDSPWEQLKHTELDDHIYVLTPQARVIALPQGATPVDFAYYLHSDLGHRCRGARVDGTMVPLNTPLKNGQTVEIVTVKQGGPSRDWLNPELHYLASNRARAKVRAWFNALELEETLAQGRILIDKTLQREGKTAVNLEELAAKLGFKTPDELYAMVAKDEFSLRHVEAVLRTDGAPPAPADDEASITKKSRATSVARGAKSGVLVVGVDSLLTQMSRCCKPAPPDLIVGFVTRGRGVSIHRQNCATFQQLAARAPGRVIQTEWGNRGEAVYPVDIHVEALDRQGLLRDISEILSREKINVTGVRTLSSKGVAKMQFTAEVSEATQLQRALILIEEVTGVLSARRK\"\n",
    "s_msa = \"VNSKTDRPAEGAGATADRAAREAAVARALECLADHAGDHQTLTGEPLVSHARGTVAILEGLRVDAPSLQAAALFLLPILATDNEKAIEPAFGGEVARLVHDVRQLLRIGAIAGMASPAEAGVTRKNEAEARRAQVEALRKMLLAFAQDIRVVLIRLASRLQSLRWLAQNKLPAPEGMARETLDIYAPLANRLGIWQLKWELEDLGFRFEDPDTYKRIARLLDEKRIEREQFIGEAIARLQQTLRDAGIHAEVSGRPKHIYSIWRKMRGKDLDFADLYDVRAFRVIVDDIKDCYTVLGFVHHMWQPIPKEFDDYISRPKTNGYKSLHTVVIGDDGRALEVQIRTREMHHFAEYGVAAHWRYKEAGSKGYAGQFSASERYDEKIAWLRQLLAWKDDAEHTVAQEDSPWEQLKHTELDDHIYVLTPQARVIALPQGATPVDFAYYLHSDLGHRCRGARVDGTMVPLNTPLKNGQTVEIVTVKQGGPSRDWLNPELHYLASNRARAKVRAWFNALELEETLAQGRILIDKTLQREGKTAVNLEELAAKLGFKTPDELYAMVAKDEFSLRHVEAVLRTDGAPPAPADDEASITKKSRATSVARGAKSGVLVVGVDSLLTQMSRCCKPAPPDLIVGFVTRGRGVSIHRQNCATFQQLAARAPGRVIQTEWGNRGEAVYPVDIHVEALDRQGLLRDISEILSREKINVTGVRTLSSKGVAKMQFTAEVSEATQLQRALILIEEVTGVLSARRK\"\n",
    "\n",
    "#s1-s2\n",
    "output_list = [li for li in difflib.ndiff(s_nucl, s_msa) if li[0] != ' ']\n",
    "print(output_list)\n",
    "\n",
    "repeat_seq = ''.join([i[-1] for i in output_list if i[0] == '-'])\n",
    "print('rep:', repeat_seq)\n",
    "\n",
    "# count diff in nucl string \n",
    "print(s_nucl.count(repeat_seq[1:]))\n",
    "print(s_msa.count(repeat_seq[1:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SAVE PROTEIN ACC FOR MAPPING\n",
    "\n",
    "with open(aln2acc, 'wb') as handle:\n",
    "    pickle.dump(g2acc, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# replace ambiguous bases with gaps!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n",
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n",
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n",
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n",
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n",
      "before {'T', 'C', 'A', 'G'}\n",
      "after  {'T', 'C', 'A', 'G'}\n"
     ]
    }
   ],
   "source": [
    "for g in g2nucl:\n",
    "    atcg = {'A', 'C', 'T', 'G'}\n",
    "    ambig = set(g2nucl[g]).difference(atcg)\n",
    "    \n",
    "    print(\"before\", set(g2nucl[g]))\n",
    "    for b in ambig:\n",
    "        g2nucl[g] = g2nucl[g].replace(b, '-')\n",
    "    print(\"after \",set(g2nucl[g]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "#str(prot_db.loc[\"WP_097388881.1\",\"GCF_002501565.1_ASM250156v1_protein\"]['nucl'])\n",
    "#for v in prot_db['GCF_001587135.1_ASM158713v1_protein'].values:\n",
    "#    print(v)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FJAT-1458 3816252\n",
      "FJAT445.F50 3816393\n",
      "FJAT15353.F50 3813270\n",
      "FJAT15304.F6 3814989\n",
      "FJAT1303.F50 3812958\n",
      "FJAT91-F8 3815583\n"
     ]
    }
   ],
   "source": [
    "msa_fasta = open(msa_out_fasta, \"w\")\n",
    "for g in g2nucl:\n",
    "    print(g, len(g2nucl[g]))\n",
    "    msa_fasta.write('>'+g+'\\n') ##+g.split('_')[2][:10]+'\\n')\n",
    "    \n",
    "    i = 0\n",
    "    for gap in g2gaps[g]:\n",
    "        if gap:\n",
    "            msa_fasta.write('-')\n",
    "        else:\n",
    "            msa_fasta.write(g2nucl[g][i])\n",
    "            i = i + 1\n",
    "    msa_fasta.write('\\n')\n",
    "msa_fasta.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert to Phylip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SeqIO.convert(msa_out_fasta, \"fasta\", msa_out_phylip, \"nexus\")\n",
    "\n",
    "nexus = open(msa_out_phylip, 'w')\n",
    "nexus.write('#NEXUS\\n')\n",
    "nexus.write('begin data;\\n')\n",
    "nexus.write('dimensions ntax='+str(len(genomes_for_og))+' nchar='+str(len(g2gaps[g]))+' ;\\n')\n",
    "nexus.write('format datatype=dna ;\\n')\n",
    "nexus.write('matrix\\n')\n",
    "\n",
    "for record in SeqIO.parse(msa_out_fasta, \"fasta\"):\n",
    "    nexus.write(record.description+'\\t'+str(record.seq)+'\\n')\n",
    "    \n",
    "nexus.write(';\\n')\n",
    "nexus.write('end;\\n')\n",
    "\n",
    "nexus.close()\n",
    "#SeqIO.write(annotated_records, msa_out_phylip, \"nexus\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export XMFA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_out_xmfa = open(msa_out_xmfa_path, 'w')\n",
    "\n",
    "for og in ogs:\n",
    "    pmsa = path_p_aln + '/' +og + '.fa'\n",
    "    for record in SeqIO.parse(pmsa, \"fasta\"):\n",
    "        #print(record.id)\n",
    "        #g_id = record.id.split('_WP')[0].replace('_1_', '.1_')\n",
    "        #p_id = record.id.split('_protein_')[-1]\n",
    "        # prokka changes \n",
    "        ##g_id = record.id.split('_')[0]\n",
    "        ##p_id = record.id.replace(record.id.split('_')[0]+'_', '')\n",
    "        \n",
    "        # prokka/ncbi genomes header structure \n",
    "        header = record.id.split('_')\n",
    "        p_id = header[-2]+'_'+header[-1]\n",
    "        g_id = strainsNames2fileNames[record.id.replace(p_id, '').replace('protein_', '')[:-1].replace('_1_', '.1_')]       \n",
    "        #print(g_id, p_id)\n",
    "                                           \n",
    "        nns = zlib.decompress(prot_db.loc[p_id, g_id]['nucl']).decode() \n",
    "        #msa_out_xmfa.write('>'+record.id.split('_')[3][:10]+'\\n')\n",
    "        #prokka\n",
    "        msa_out_xmfa.write('>'+g_id+'\\n')\n",
    "        \n",
    "        atcg = {'A', 'T', 'G', 'C','-'}\n",
    "        ambig = set(nns).difference(atcg)\n",
    "        #print(ambig)\n",
    "        for b in ambig:\n",
    "            nns = str(nns).replace(b, '-')\n",
    "        #print('aaaaaa',  nns)\n",
    "        \n",
    "        gaps_v = gaps2tab.loc[og, g_id]\n",
    "        i = 0\n",
    "        for gap in gaps_v:\n",
    "            if gap:\n",
    "                msa_out_xmfa.write('-')\n",
    "            else:\n",
    "                msa_out_xmfa.write(nns[i])\n",
    "                i = i + 1\n",
    "        msa_out_xmfa.write('\\n')\n",
    "    msa_out_xmfa.write('=' + '\\n')\n",
    "msa_out_xmfa.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test the position mapping\n",
    "+ grab sequences from the original gb files using g2pos, g2strand and g2chr\n",
    "+ load exported core OG nucleotide fasta file\n",
    "+ compare the sequences \n",
    "\n",
    "(note: reverse compliment!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DEBUG:\n",
    "    # for the compliment strand mapping\n",
    "    from Bio.Seq import Seq\n",
    "    Seq(\"A\").complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome: FJAT1303.F50\n",
      "Chr NZ_CP052126.1 with 7264 features\n",
      "Chr NZ_CP052127.1 with 3123 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n",
      "Genome: FJAT445.F50\n",
      "Chr NZ_CP052076.1 with 6856 features\n",
      "Chr NZ_CP052077.1 with 3177 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n",
      "Genome: FJAT15304.F6\n",
      "Chr NZ_CP052096.1 with 6822 features\n",
      "Chr NZ_CP052097.1 with 3221 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n",
      "Genome: FJAT-1458\n",
      "Chr NZ_CP016554.1 with 7588 features\n",
      "Chr NZ_CP016555.1 with 3271 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n",
      "Genome: FJAT15353.F50\n",
      "Chr NZ_CP052086.1 with 7262 features\n",
      "Chr NZ_CP052087.1 with 3127 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n",
      "Genome: FJAT91-F8\n",
      "Chr NZ_CP056085.1 with 7282 features\n",
      "Chr NZ_CP056086.1 with 3165 features\n",
      "[]\n",
      "[]\n",
      "[]\n",
      "0 0\n",
      "Core nucl exported: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if DEBUG:\n",
    "    import os\n",
    "    gbks = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/inputs/panX/'+ids\n",
    "    ccc=0\n",
    "\n",
    "    test_fasta = {}\n",
    "    for entry in os.scandir(gbks):\n",
    "        #if ccc>0:\n",
    "        #    break\n",
    "        ccc = ccc+1\n",
    "        if (entry.path.endswith(\".gbk\") and entry.is_file()):\n",
    "            genome = os.path.basename(entry).replace('.gbk', '').replace('_genomic', '')\n",
    "            genome = strainsNames2fileNames[genome]\n",
    "            test_fasta[genome] = []\n",
    "            print('Genome:', genome)\n",
    "\n",
    "            # map chr id to the sequence\n",
    "            chr_seq = {}\n",
    "            for record in SeqIO.parse(entry.path, \"gb\"):\n",
    "                print('Chr',record.id, 'with', len(record.features), 'features')\n",
    "                chr_seq[record.id] = record.seq\n",
    "\n",
    "            # get corresponding mapping (from above)\n",
    "            mapping_chr = g2chr[genome]\n",
    "            mapping_pos = g2pos[genome]\n",
    "            mapping_str = g2strand[genome]\n",
    "            \n",
    "            print(g2chr[genome][:10])\n",
    "            print(g2pos[genome][:10])\n",
    "            print(g2strand[genome][:10])\n",
    "            \n",
    "            print(len(mapping_chr), len(mapping_pos))\n",
    "            assert(len(mapping_chr) == len(mapping_pos))\n",
    "\n",
    "            # iterate through the original positions \n",
    "            for i in range(len(mapping_chr)):\n",
    "                if mapping_str[i] == 1:\n",
    "                    test_fasta[genome].append(chr_seq[mapping_chr[i]][mapping_pos[i]])\n",
    "                else:\n",
    "                    compl = str(Seq(chr_seq[mapping_chr[i]][mapping_pos[i]]).complement())\n",
    "                    test_fasta[genome].append(compl)\n",
    "            test_fasta[genome] = \"\".join(test_fasta[genome])\n",
    "            print(\"Core nucl exported:\", len(test_fasta[genome]))\n",
    "\n",
    "            print()\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FJAT-1458 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n",
      "FJAT445.F50 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n",
      "FJAT15353.F50 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n",
      "FJAT15304.F6 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n",
      "FJAT1303.F50 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n",
      "FJAT91-F8 3825198\n",
      "test length 3825198\n",
      "Test: False\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if DEBUG:\n",
    "\n",
    "    for record in SeqIO.parse(msa_out_fasta, \"fasta\"):\n",
    "        print(record.id, len(record.seq))\n",
    "        print('test length', len(record.seq))\n",
    "        print('Test:', test_fasta[record.id] == str(record.seq))\n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "f = \"/Users/devseeva/Desktop/work/pan-genome-visualization/data/ncbi-rso/metainfo.tsv\"\n",
    "x = pd.read_csv(f, sep='\\t')  \n",
    "phys = []\n",
    "for a in x[\"accession\"]:\n",
    "    phys.append(a.split(\".1_\")[1][:10])\n",
    "x[\"phy_id\"] = phys\n",
    "x.to_csv(\"/Users/devseeva/Desktop/work/sm_workflow/snakefiles/inputs/phy4_test_metadata/phy4_test_cleaned_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
