{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "communist-short",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import zlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "single-mining",
   "metadata": {},
   "source": [
    "# Remove recombinations from MSA\n",
    "\n",
    "+ input: MSA (fa, xmfa), tree, predicted importations \n",
    "+ output 1: MSA without recombination\n",
    "    + fasta\n",
    "    + phylip\n",
    "+ output 2: table of protein ids assigned to importations \n",
    "\n",
    "\n",
    "# Results for the test sets\n",
    "\n",
    "## MSA lengths\n",
    "+ rso-test (fjat) = 3,825,198 (xmfa: 7,666,198)\n",
    "+ phy2 = 2,987,190 (xmfa: 6,086,190)\n",
    "+ phy4 = 3,219,942\n",
    "+ phy3 = 3,459,843 (xmfa: 6,947,843)\n",
    "\n",
    "(for ASM359060v in phy2 dataset the summed length is 116,425)\n",
    "\n",
    "## Summed recombinations length\n",
    "+ rso = 92,925\n",
    "+ phy2 = 290,218\n",
    "+ phy4 = 58,063\n",
    "+ phy3 = 120,259\n",
    "\n",
    "(phy4 strain PSI07 is from Indonesia! very different)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "temporal-measure",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = 'korea'#'rso_test'\n",
    "\n",
    "recombis = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_cfml/rso_test_cfml.importation_status.txt'\n",
    "path2tree = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_cfml/rso_test_cfml.labelled_tree.newick'\n",
    "xm_msa_in = \"/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_concat.xmfa\"\n",
    "fa_msa_in = \"/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_concat.fasta\"\n",
    "accDB = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/MSA_ACC_DB_rso_test.pkl'\n",
    "protDB = '../'+SAMPLE+'_outputs/GENOME_PROTEIN_DB_'+SAMPLE+'.pkl'\n",
    "# out\n",
    "fa_msa_out = \"/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_concat_no_RR.fasta\"\n",
    "ph_msa_out = \"/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_concat_no_RR.phylip\"\n",
    "updated_recombis = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_recombis2prot.csv'\n",
    "mult_recombis = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_recombis_mult_genes.txt'\n",
    "top_recombis = '/Users/devseeva/Desktop/work/sm_workflow/geneTrack/rso_test_outputs/rso_test_recombis_top_genes.txt'\n",
    "\n",
    "\n",
    "# in\n",
    "recombis = recombis.replace(\"rso_test\", SAMPLE)\n",
    "path2tree = path2tree.replace(\"rso_test\", SAMPLE)\n",
    "xm_msa_in = xm_msa_in.replace(\"rso_test\", SAMPLE)\n",
    "fa_msa_in = fa_msa_in.replace(\"rso_test\", SAMPLE)\n",
    "accDB = accDB.replace(\"rso_test\", SAMPLE)\n",
    "protDB = protDB.replace(\"rso_test\", SAMPLE)\n",
    "# out\n",
    "fa_msa_in = fa_msa_in.replace(\"rso_test\", SAMPLE)\n",
    "ph_msa_out = ph_msa_out.replace(\"rso_test\", SAMPLE)\n",
    "updated_recombis = updated_recombis.replace(\"rso_test\", SAMPLE)\n",
    "mult_recombis = mult_recombis.replace(\"rso_test\", SAMPLE)\n",
    "top_recombis = top_recombis.replace(\"rso_test\", SAMPLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37869b39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# in\n",
    "recombis = snakemake.params[\"recombis\"]\n",
    "path2tree = snakemake.params[\"tree\"]\n",
    "xm_msa_in = snakemake.input[\"xm_msa_in\"]\n",
    "fa_msa_in = snakemake.input[\"fa_msa_in\"]\n",
    "accDB = snakemake.input[\"accDB\"]\n",
    "protDB = snakemake.input[\"protDB\"]\n",
    "# out\n",
    "fa_msa_out = snakemake.output[\"fa_msa_out\"]\n",
    "ph_msa_out = snakemake.output[\"ph_msa_out\"]\n",
    "updated_recombis = snakemake.output[\"updated_recombis\"]\n",
    "mult_recombis = snakemake.output[\"mult_recombis\"]\n",
    "top_recombis = snakemake.output[\"top_recombis\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "damaged-alberta",
   "metadata": {},
   "source": [
    "# Replace the inner nodes with the leaves "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "temporal-pilot",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Beg</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pe13</td>\n",
       "      <td>1951157</td>\n",
       "      <td>1951163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pe13</td>\n",
       "      <td>3007423</td>\n",
       "      <td>3007452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pe13</td>\n",
       "      <td>3008403</td>\n",
       "      <td>3008431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pe13</td>\n",
       "      <td>3008682</td>\n",
       "      <td>3008688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pe13</td>\n",
       "      <td>5281084</td>\n",
       "      <td>5281091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>NODE_13</td>\n",
       "      <td>6402045</td>\n",
       "      <td>6406018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>678</th>\n",
       "      <td>NODE_13</td>\n",
       "      <td>6406361</td>\n",
       "      <td>6410314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>679</th>\n",
       "      <td>NODE_13</td>\n",
       "      <td>6410572</td>\n",
       "      <td>6412423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680</th>\n",
       "      <td>NODE_13</td>\n",
       "      <td>6413081</td>\n",
       "      <td>6413086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>681</th>\n",
       "      <td>NODE_13</td>\n",
       "      <td>6414181</td>\n",
       "      <td>6414184</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>682 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Node      Beg      End\n",
       "0       pe13  1951157  1951163\n",
       "1       pe13  3007423  3007452\n",
       "2       pe13  3008403  3008431\n",
       "3       pe13  3008682  3008688\n",
       "4       pe13  5281084  5281091\n",
       "..       ...      ...      ...\n",
       "677  NODE_13  6402045  6406018\n",
       "678  NODE_13  6406361  6410314\n",
       "679  NODE_13  6410572  6412423\n",
       "680  NODE_13  6413081  6413086\n",
       "681  NODE_13  6414181  6414184\n",
       "\n",
       "[682 rows x 3 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imp_stat = pd.read_csv(recombis, sep='\\t')\n",
    "imp_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "accompanied-lightning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NODE_15': ['pe26', 'pe61', 'to53', 'pe15', 'pe30', 'pe9', 'to28', 'pe13'],\n",
       " 'NODE_14': ['pe26', 'pe61', 'to53', 'pe15', 'pe30', 'pe9', 'to28'],\n",
       " 'NODE_13': ['pe26', 'pe61', 'to53', 'pe15', 'pe30', 'pe9'],\n",
       " 'NODE_12': ['pe26', 'pe61', 'to53', 'pe15', 'pe30'],\n",
       " 'NODE_10': ['pe26', 'pe61', 'to53'],\n",
       " 'NODE_11': ['pe15', 'pe30'],\n",
       " 'NODE_9': ['pe26', 'pe61']}"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for tree in Phylo.parse(path2tree, 'newick'):\n",
    "    parents = {}\n",
    "    # iterate through inner nodes \n",
    "    for clade in tree.get_nonterminals(order=\"level\"):\n",
    "        parents[clade.name] = []\n",
    "        # get leaves for each inner node\n",
    "        for child in clade.get_terminals():\n",
    "            parents[clade.name].append(child.name)\n",
    "        #parents[clade.name] = \";\".join(parents[clade.name])\n",
    "parents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "empty-distributor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NODE_15 has 0 recombinations and 8 children\n",
      "\n",
      "NODE_14 has 0 recombinations and 7 children\n",
      "\n",
      "NODE_13 has 5 recombinations and 6 children\n",
      "\n",
      "NODE_12 has 0 recombinations and 5 children\n",
      "\n",
      "NODE_10 has 50 recombinations and 3 children\n",
      "\n",
      "NODE_11 has 6 recombinations and 2 children\n",
      "\n",
      "NODE_9 has 199 recombinations and 2 children\n",
      "\n",
      "DF length after parent nodes expantion = 1012 ; Must be = 1012\n"
     ]
    }
   ],
   "source": [
    "new_len = len(imp_stat)\n",
    "\n",
    "for p in parents:\n",
    "\n",
    "    is_node = imp_stat[\"Node\"].str.startswith(p)\n",
    "    num_child = len(parents[p]) \n",
    "    print(p,'has',len(imp_stat[is_node]),'recombinations and',num_child,'children')\n",
    "    new_len = new_len + len(imp_stat[is_node]) * (num_child - 1)\n",
    "    \n",
    "    #imp_stat = imp_stat.append([imp_stat[is_node]]*(num_child-1), ignore_index=True)\n",
    "    print()\n",
    "    \n",
    "    # for each child: duplicate the parent row\n",
    "    for child in parents[p]:       \n",
    "        node_imp = imp_stat[is_node].replace({p:child}) \n",
    "        imp_stat = imp_stat.append([node_imp], ignore_index=True)\n",
    "        is_node = imp_stat[\"Node\"].str.startswith(p)\n",
    "\n",
    "# remove the remaining parent rows\n",
    "\n",
    "imp_stat = imp_stat.loc[~imp_stat[\"Node\"].str.contains(\"NODE_\")]\n",
    "print('DF length after parent nodes expantion =', len(imp_stat), '; Must be =', new_len)\n",
    "assert(new_len == len(imp_stat))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "steady-mystery",
   "metadata": {},
   "source": [
    "# Sort and sum the importations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "interior-spine",
   "metadata": {},
   "outputs": [],
   "source": [
    "imp_stat = imp_stat.sort_values(by=['Beg', 'End'])\n",
    "#imp_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "guilty-valve",
   "metadata": {},
   "outputs": [],
   "source": [
    "beg = imp_stat[\"Beg\"].values.copy()\n",
    "end = imp_stat[\"End\"].values.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "interpreted-metabolism",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Summed recombination length: 53735\n"
     ]
    }
   ],
   "source": [
    "ranges2del = []\n",
    "for i in range(len(imp_stat)):\n",
    "    ranges2del.extend(range(beg[i]-1, end[i]))\n",
    "ranges2del = list(dict.fromkeys(ranges2del))\n",
    "print('Summed recombination length:', len(ranges2del))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "located-sharp",
   "metadata": {},
   "source": [
    "# Fasta vs. XMFA positions mapping list\n",
    "\n",
    "* list of updated positions \n",
    "* index represent Fasta sequence positions \n",
    "* values represent XMFA sequence positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "substantial-mercury",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found genes MSAs / LCBs: 3996\n"
     ]
    }
   ],
   "source": [
    "# parse the genes MSAs into vector of the alignments' lengths\n",
    "aln_length = []\n",
    "aln = open(xm_msa_in, 'r')\n",
    "\n",
    "for l in aln:\n",
    "    \n",
    "    if l.startswith('>'):\n",
    "        c = 0\n",
    "    elif l.startswith('='):\n",
    "        aln_length.append(c-1) # minus one char for the \\n symbol\n",
    "    else:\n",
    "        c = c + len(l)\n",
    "        \n",
    "print('Found genes MSAs / LCBs:', len(aln_length))    \n",
    "aln.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "atmospheric-judges",
   "metadata": {},
   "outputs": [],
   "source": [
    "# insert 1000 count after each gene MSA\n",
    "index_mapping = list(range(0,aln_length[0]))\n",
    "sum_lengths = aln_length[0]\n",
    "\n",
    "for i in range(1,len(aln_length)):  \n",
    "    index_mapping.extend(range(sum_lengths + i*1000, sum_lengths + aln_length[i] + i*1000))\n",
    "    sum_lengths = sum_lengths + aln_length[i]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "polish-raleigh",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fasta length 4089060 \n",
      "XMFA length 8084060\n"
     ]
    }
   ],
   "source": [
    "print('Fasta length', sum(aln_length), '\\nXMFA length', sum(aln_length)+(len(aln_length)-1)*1000)\n",
    "# index = fasta file length\n",
    "assert(sum(aln_length) == len(index_mapping))\n",
    "# values = xmfa file length\n",
    "assert(sum(aln_length)+(len(aln_length)-1)*1000 == index_mapping[-1]+1) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "improved-columbia",
   "metadata": {},
   "source": [
    "# Get the fasta position of each insertion from the xmfa position\n",
    "\n",
    "+ use ranges to improve the running time of .index(i, start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "smoking-excellence",
   "metadata": {},
   "outputs": [],
   "source": [
    "fa_beg = []\n",
    "fa_end = []\n",
    "previous_beg = 0\n",
    "previous_end = 0\n",
    "L=len(index_mapping)\n",
    "\n",
    "for r_i in range(len(imp_stat)):\n",
    "    \n",
    "    # -1 for 0-start indexing\n",
    "    r_beg = beg[r_i] - 1\n",
    "    r_end = end[r_i] - 1\n",
    "    #print(r_beg, r_end)\n",
    "    \n",
    "    # USE PREVIOUS_BEG FOR BOTH BECAUSE OF THE SORTING ISSUE\n",
    "    r_beg_index = index_mapping.index(r_beg, previous_beg, L)#r_beg+1)\n",
    "    r_end_index = index_mapping.index(r_end, previous_beg, L)#r_end+1)\n",
    "    \n",
    "    previous_beg = r_beg_index\n",
    "    previous_end = r_end_index\n",
    "    \n",
    "    fa_beg.append(r_beg_index)\n",
    "    fa_end.append(r_end_index)\n",
    "    \n",
    "    #print(r_i,\"xmfa - fa:\",r_beg_index, r_end_index)\n",
    "\n",
    "imp_stat['FA Beg'] = fa_beg\n",
    "imp_stat['FA End'] = fa_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "minute-province",
   "metadata": {},
   "outputs": [],
   "source": [
    "imp_stat['debug_len_must'] = imp_stat['End'] - imp_stat['Beg']\n",
    "imp_stat[\"debug_len_is\"] = imp_stat['FA End'] - imp_stat['FA Beg']\n",
    "#assert(len(imp_stat.loc[imp_stat['debug_len_must'] != imp_stat['debug_len_is']]) == 0)\n",
    "#imp_stat\n",
    "imp_stat.loc[imp_stat['debug_len_must'] != imp_stat['debug_len_is']].to_csv(mult_recombis, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acoustic-upper",
   "metadata": {},
   "source": [
    "# Delete the recombinations from the fasta file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "adequate-slovak",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome pe13 has 360 recombinations\n",
      "Number of gaps before 33132\n",
      "Number of gaps after 33401\n",
      "\n",
      "Genome pe15 has 10651 recombinations\n",
      "Number of gaps before 33003\n",
      "Number of gaps after 43207\n",
      "\n",
      "Genome pe26 has 38512 recombinations\n",
      "Number of gaps before 44475\n",
      "Number of gaps after 71541\n",
      "\n",
      "Genome pe30 has 10543 recombinations\n",
      "Number of gaps before 33639\n",
      "Number of gaps after 43753\n",
      "\n",
      "Genome pe61 has 38959 recombinations\n",
      "Number of gaps before 39042\n",
      "Number of gaps after 76904\n",
      "\n",
      "Genome pe9 has 9857 recombinations\n",
      "Number of gaps before 34092\n",
      "Number of gaps after 43697\n",
      "\n",
      "Genome to28 has 401 recombinations\n",
      "Number of gaps before 29634\n",
      "Number of gaps after 29967\n",
      "\n",
      "Genome to53 has 20680 recombinations\n",
      "Number of gaps before 46062\n",
      "Number of gaps after 65889\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fasta_out = open(fa_msa_out, 'w')\n",
    "num_of_strains = 0\n",
    "for record in SeqIO.parse(fa_msa_in, 'fasta'):\n",
    "    num_of_strains = num_of_strains + 1\n",
    "    fasta_out.write('>' + record.id + '\\n')\n",
    "    genome_imp_stat = imp_stat.loc[imp_stat['Node'] == record.id]\n",
    "    beg = genome_imp_stat[\"FA Beg\"].values.copy()\n",
    "    end = genome_imp_stat[\"FA End\"].values.copy()\n",
    "    \n",
    "    ranges2del = []\n",
    "    for i in range(len(genome_imp_stat)):\n",
    "        ranges2del.extend(range(beg[i]-1, end[i]))\n",
    "    ranges2del = list(dict.fromkeys(ranges2del))\n",
    "    \n",
    "    print('Genome',record.id,'has',len(ranges2del),'recombinations')\n",
    "    #print(ranges2del)\n",
    "    \n",
    "    print('Number of gaps before',record.seq.count('-'))\n",
    "    seq = list(record.seq)\n",
    "    for to_del in ranges2del:\n",
    "        seq[to_del] = '-'\n",
    "    seq = ''.join(seq)\n",
    "    print('Number of gaps after',seq.count('-'))\n",
    "    print()\n",
    "    fasta_out.write(seq + '\\n')\n",
    "    \n",
    "fasta_out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "green-digest",
   "metadata": {},
   "source": [
    "# Convert to Phylip TODO:length!!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a47855e4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "incredible-advice",
   "metadata": {},
   "outputs": [],
   "source": [
    "#SeqIO.convert(fa_msa_out, \"fasta\", ph_msa_out, \"phylip\")\n",
    "\n",
    "nexus = open(ph_msa_out, 'w')\n",
    "nexus.write('#NEXUS\\n')\n",
    "nexus.write('begin data;\\n')\n",
    "nexus.write('dimensions ntax='+str(num_of_strains)+' nchar='+str(sum(aln_length))+' ;\\n')\n",
    "nexus.write('format datatype=dna ;\\n')\n",
    "nexus.write('matrix\\n')\n",
    "\n",
    "for record in SeqIO.parse(fa_msa_out, \"fasta\"):\n",
    "    nexus.write(record.description+'\\t'+str(record.seq)+'\\n')\n",
    "    \n",
    "nexus.write(';\\n')\n",
    "nexus.write('end;\\n')\n",
    "\n",
    "nexus.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "coordinate-resident",
   "metadata": {},
   "source": [
    "# Expand and export the importation_status table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "decent-processing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['pe13', 'pe15', 'pe26', 'pe30', 'pe61', 'pe9', 'to28', 'to53'])"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read Accession Database\n",
    "\n",
    "with open(accDB, 'rb') as handle:\n",
    "    acc = pickle.load(handle)\n",
    "acc.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "attended-large",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>to53</th>\n",
       "      <th>pe15</th>\n",
       "      <th>pe13</th>\n",
       "      <th>pe61</th>\n",
       "      <th>to28</th>\n",
       "      <th>pe9</th>\n",
       "      <th>pe26</th>\n",
       "      <th>pe30</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GDEOEPAP_00001</th>\n",
       "      <td>{'chr': '1', 'location': (30, 31, 32, 33, 34, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDEOEPAP_00002</th>\n",
       "      <td>{'chr': '1', 'location': (1876, 1877, 1878, 18...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDEOEPAP_00003</th>\n",
       "      <td>{'chr': '1', 'location': (3128, 3129, 3130, 31...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDEOEPAP_00004</th>\n",
       "      <td>{'chr': '1', 'location': (7139, 7138, 7137, 71...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDEOEPAP_00005</th>\n",
       "      <td>{'chr': '1', 'location': (8581, 8582, 8583, 85...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EENJCLCO_05336</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'cluster_012', 'location': (8609, 8610...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EENJCLCO_05337</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'cluster_012', 'location': (10200, 101...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EENJCLCO_05338</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'cluster_012', 'location': (11858, 118...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EENJCLCO_05339</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'cluster_012', 'location': (12330, 123...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EENJCLCO_05340</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>{'chr': 'cluster_012', 'location': (12668, 126...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41214 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             to53 pe15 pe13  \\\n",
       "GDEOEPAP_00001  {'chr': '1', 'location': (30, 31, 32, 33, 34, ...  NaN  NaN   \n",
       "GDEOEPAP_00002  {'chr': '1', 'location': (1876, 1877, 1878, 18...  NaN  NaN   \n",
       "GDEOEPAP_00003  {'chr': '1', 'location': (3128, 3129, 3130, 31...  NaN  NaN   \n",
       "GDEOEPAP_00004  {'chr': '1', 'location': (7139, 7138, 7137, 71...  NaN  NaN   \n",
       "GDEOEPAP_00005  {'chr': '1', 'location': (8581, 8582, 8583, 85...  NaN  NaN   \n",
       "...                                                           ...  ...  ...   \n",
       "EENJCLCO_05336                                                NaN  NaN  NaN   \n",
       "EENJCLCO_05337                                                NaN  NaN  NaN   \n",
       "EENJCLCO_05338                                                NaN  NaN  NaN   \n",
       "EENJCLCO_05339                                                NaN  NaN  NaN   \n",
       "EENJCLCO_05340                                                NaN  NaN  NaN   \n",
       "\n",
       "               pe61 to28  pe9 pe26  \\\n",
       "GDEOEPAP_00001  NaN  NaN  NaN  NaN   \n",
       "GDEOEPAP_00002  NaN  NaN  NaN  NaN   \n",
       "GDEOEPAP_00003  NaN  NaN  NaN  NaN   \n",
       "GDEOEPAP_00004  NaN  NaN  NaN  NaN   \n",
       "GDEOEPAP_00005  NaN  NaN  NaN  NaN   \n",
       "...             ...  ...  ...  ...   \n",
       "EENJCLCO_05336  NaN  NaN  NaN  NaN   \n",
       "EENJCLCO_05337  NaN  NaN  NaN  NaN   \n",
       "EENJCLCO_05338  NaN  NaN  NaN  NaN   \n",
       "EENJCLCO_05339  NaN  NaN  NaN  NaN   \n",
       "EENJCLCO_05340  NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                             pe30  \n",
       "GDEOEPAP_00001                                                NaN  \n",
       "GDEOEPAP_00002                                                NaN  \n",
       "GDEOEPAP_00003                                                NaN  \n",
       "GDEOEPAP_00004                                                NaN  \n",
       "GDEOEPAP_00005                                                NaN  \n",
       "...                                                           ...  \n",
       "EENJCLCO_05336  {'chr': 'cluster_012', 'location': (8609, 8610...  \n",
       "EENJCLCO_05337  {'chr': 'cluster_012', 'location': (10200, 101...  \n",
       "EENJCLCO_05338  {'chr': 'cluster_012', 'location': (11858, 118...  \n",
       "EENJCLCO_05339  {'chr': 'cluster_012', 'location': (12330, 123...  \n",
       "EENJCLCO_05340  {'chr': 'cluster_012', 'location': (12668, 126...  \n",
       "\n",
       "[41214 rows x 8 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read Protein Database\n",
    "\n",
    "prot_db = pd.read_pickle(protDB)\n",
    "prot_db = prot_db.rename(columns=lambda x: str(x).split('$')[0])\n",
    "prot_db"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specified-egyptian",
   "metadata": {},
   "source": [
    "for k in acc.keys():\n",
    "    g2ranges = imp_stat.loc[imp_stat['Node'] == k]\n",
    "    \n",
    "    g_beg = g2ranges[\"FA Beg\"].values.copy()\n",
    "    g_end = g2ranges[\"FA End\"].values.copy()\n",
    "    \n",
    "    print(k)\n",
    "    for i in range(len(g_beg)):\n",
    "        print(set(acc[k][g_beg[i]:g_end[i]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "a24c6ac7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['pe13', 'pe15', 'pe26', 'pe30', 'pe61', 'pe9', 'to28', 'to53'])"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acc.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "european-heavy",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign improtations to proteins\n",
    "\n",
    "g_beg = imp_stat[\"FA Beg\"].values.copy()\n",
    "g_end = imp_stat[\"FA End\"].values.copy()\n",
    "g_node = imp_stat[\"Node\"].values.copy()\n",
    "\n",
    "import_acc = []\n",
    "import_products = []\n",
    "for i in range(len(g_beg)):\n",
    "    import_acc.append(set(acc[g_node[i]][g_beg[i]:g_end[i]]))\n",
    "    \n",
    "    products = []\n",
    "    for a in import_acc[-1]:\n",
    "        products.append(zlib.decompress(prot_db.loc[a, g_node[i]]['product']).decode())\n",
    "    import_products.append(products)\n",
    "    \n",
    "imp_stat['Prot-ACC'] = import_acc\n",
    "imp_stat['Product'] = import_products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "administrative-petersburg",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Node</th>\n",
       "      <th>Beg</th>\n",
       "      <th>End</th>\n",
       "      <th>FA Beg</th>\n",
       "      <th>FA End</th>\n",
       "      <th>debug_len_must</th>\n",
       "      <th>debug_len_is</th>\n",
       "      <th>Prot-ACC</th>\n",
       "      <th>Product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>276</th>\n",
       "      <td>to53</td>\n",
       "      <td>6423</td>\n",
       "      <td>6437</td>\n",
       "      <td>4422</td>\n",
       "      <td>4436</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>{GDEOEPAP_00003}</td>\n",
       "      <td>[[DNA gyrase subunit B]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>pe26</td>\n",
       "      <td>6671</td>\n",
       "      <td>6680</td>\n",
       "      <td>4670</td>\n",
       "      <td>4679</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>{NCAEKIBJ_00003}</td>\n",
       "      <td>[[DNA gyrase subunit B]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>pe61</td>\n",
       "      <td>6671</td>\n",
       "      <td>6680</td>\n",
       "      <td>4670</td>\n",
       "      <td>4679</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>{KPJNPNMP_00003}</td>\n",
       "      <td>[[DNA gyrase subunit B]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>to53</td>\n",
       "      <td>6671</td>\n",
       "      <td>6680</td>\n",
       "      <td>4670</td>\n",
       "      <td>4679</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>{GDEOEPAP_00003}</td>\n",
       "      <td>[[DNA gyrase subunit B]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>874</th>\n",
       "      <td>pe26</td>\n",
       "      <td>6839</td>\n",
       "      <td>6941</td>\n",
       "      <td>4838</td>\n",
       "      <td>4940</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>{NCAEKIBJ_00003}</td>\n",
       "      <td>[[DNA gyrase subunit B]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pe13</td>\n",
       "      <td>8024479</td>\n",
       "      <td>8024551</td>\n",
       "      <td>4063478</td>\n",
       "      <td>4063550</td>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>{AIELFPIA_05293}</td>\n",
       "      <td>[[hypothetical protein]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1071</th>\n",
       "      <td>pe26</td>\n",
       "      <td>8025703</td>\n",
       "      <td>8025774</td>\n",
       "      <td>4063702</td>\n",
       "      <td>4063773</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>{NCAEKIBJ_05096}</td>\n",
       "      <td>[[hypothetical protein]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1270</th>\n",
       "      <td>pe61</td>\n",
       "      <td>8025703</td>\n",
       "      <td>8025774</td>\n",
       "      <td>4063702</td>\n",
       "      <td>4063773</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>{KPJNPNMP_05153}</td>\n",
       "      <td>[[hypothetical protein]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1072</th>\n",
       "      <td>pe26</td>\n",
       "      <td>8072284</td>\n",
       "      <td>8072490</td>\n",
       "      <td>4082283</td>\n",
       "      <td>4082489</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>{NCAEKIBJ_05125}</td>\n",
       "      <td>[[hypothetical protein]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1271</th>\n",
       "      <td>pe61</td>\n",
       "      <td>8072284</td>\n",
       "      <td>8072490</td>\n",
       "      <td>4082283</td>\n",
       "      <td>4082489</td>\n",
       "      <td>206</td>\n",
       "      <td>206</td>\n",
       "      <td>{KPJNPNMP_05182}</td>\n",
       "      <td>[[hypothetical protein]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1012 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Node      Beg      End   FA Beg   FA End  debug_len_must  debug_len_is  \\\n",
       "276   to53     6423     6437     4422     4436              14            14   \n",
       "712   pe26     6671     6680     4670     4679               9             9   \n",
       "762   pe61     6671     6680     4670     4679               9             9   \n",
       "812   to53     6671     6680     4670     4679               9             9   \n",
       "874   pe26     6839     6941     4838     4940             102           102   \n",
       "...    ...      ...      ...      ...      ...             ...           ...   \n",
       "9     pe13  8024479  8024551  4063478  4063550              72            72   \n",
       "1071  pe26  8025703  8025774  4063702  4063773              71            71   \n",
       "1270  pe61  8025703  8025774  4063702  4063773              71            71   \n",
       "1072  pe26  8072284  8072490  4082283  4082489             206           206   \n",
       "1271  pe61  8072284  8072490  4082283  4082489             206           206   \n",
       "\n",
       "              Prot-ACC                   Product  \n",
       "276   {GDEOEPAP_00003}  [[DNA gyrase subunit B]]  \n",
       "712   {NCAEKIBJ_00003}  [[DNA gyrase subunit B]]  \n",
       "762   {KPJNPNMP_00003}  [[DNA gyrase subunit B]]  \n",
       "812   {GDEOEPAP_00003}  [[DNA gyrase subunit B]]  \n",
       "874   {NCAEKIBJ_00003}  [[DNA gyrase subunit B]]  \n",
       "...                ...                       ...  \n",
       "9     {AIELFPIA_05293}  [[hypothetical protein]]  \n",
       "1071  {NCAEKIBJ_05096}  [[hypothetical protein]]  \n",
       "1270  {KPJNPNMP_05153}  [[hypothetical protein]]  \n",
       "1072  {NCAEKIBJ_05125}  [[hypothetical protein]]  \n",
       "1271  {KPJNPNMP_05182}  [[hypothetical protein]]  \n",
       "\n",
       "[1012 rows x 9 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imp_stat.to_csv(updated_recombis, index=False)\n",
    "imp_stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "63642e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "imp_stat = pd.read_csv(updated_recombis)\n",
    "imp_stat = imp_stat.drop(['Node'], axis=1).drop_duplicates()\n",
    "imp_stat['Product'].value_counts(normalize=True)[:50].to_csv(top_recombis)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "planned-incentive",
   "metadata": {},
   "source": [
    "# TODO {'WP_071623447.1'} seq length is less than improtation length?\n",
    "# TODO enrichment with probabilities \n",
    "e.g. numOf(T3 effectors with recombinations)/numOf(all T3 effectors in the genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "southern-antigua",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prot_db.loc[\"WP_071507848.1\",\"ASM1330693\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comparable-canvas",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
